---
title: "CC Singlecell Reproducible Analysis"
author: "Andrew Willems and Tian Hong"
date: "5/25/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Reproducible document for all CC Singlecell analysis

This document encompasses all steps, code, and data used for the CC Singlecell paper

## Step 0: Install needed packages
```{r install needed packages}
bioc_pkgs <- c(
  "BiocParallel", "biomaRt", "DESeq2", "DEsingle", "edgeR", "scDD", "scde",
  "scran", "SingleCellExperiment", "SummarizedExperiment",
  "TCGAbiolinks", "zinbwave", "BiocGenerics", "DelayedArray", "DelayedMatrixStats",
  "limma", "lme4", "S4Vectors", "SingleCellExperiment",
  "SummarizedExperiment", "batchelor", "HDF5Array",
  "terra", "ggrastr"
)
BiocManager::install(bioc_pkgs)
install.packages(c(
  "data.table", "extrafont", "ggbreak", "ggpubr", "igraph", "tidyverse", "gt", "phateR", "survival",
  "survminer", "svglite", "viridis", "devtools", "RColorBrewer", "Rmagic", "scales"
))

devtools::install_github("cole-trapnell-lab/monocle3")
```



## Step 1: Loading needed packages
```{r loading needed packages, message=FALSE, echo=FALSE}
# Loading needed packages
library(BiocParallel)
library(biomaRt)
library(data.table)
library(DESeq2)
library(DEsingle)
library(edgeR)
library(extrafont)
library(ggbreak)
library(ggplot2)
library(ggpubr)
library(grid)
library(gt)
library(igraph)
library(monocle3)
library(phateR)
library(RColorBrewer)
library(Rmagic)
library(scales)
library(scDD)
library(scde)
library(scran)
library(SingleCellExperiment)
library(SummarizedExperiment)
library(survival)
library(survminer)
library(svglite)
library(TCGAbiolinks)
library(tidyverse)
library(viridis)
library(zinbwave)
```

```{r arial font loaded}
suppressMessages(font_import())
"Arial" %in% fonts()
```



## Step 2: Loading in the functions we need from their separate files
```{r loading needed functions}
# Loading the needed functions from their respective files
source("Data/Reproducible-results/Rmarkdown/Code/cell_dataset_builder.R")
source("Data/Reproducible-results/Rmarkdown/Code/cox_model.R")
source("Data/Reproducible-results/Rmarkdown/Code/mad_calculator.R")
source("Data/Reproducible-results/Rmarkdown/Code/magic_denoiser.R")
source("Data/Reproducible-results/Rmarkdown/Code/mirna_calculator_original.R")
source("Data/Reproducible-results/Rmarkdown/Code/model_optimizer.R")
source("Data/Reproducible-results/Rmarkdown/Code/risk_score_calculator.R")
source("Data/Reproducible-results/Rmarkdown/Code/switchde_calculator.R")
```


## Step 3: Loading in the single-cell data
```{r loading single-cell colorectal cancer file}
# Loading single-cell data that is used for both colon and rectal cancer data sets
cc_tumor_fpkm <- read.csv("Data/Reproducible-results/Rmarkdown/Data/Single-cell-data/FPKM/GSE81861_CRC_tumor_all_cells_FPKM.csv")
head(cc_tumor_fpkm[, 1:3])
```

## Step 4: Doing some pre-processing to clean up the gene names
```{r pre-processing colorectal cancer file}
# Processing the single-cell data to clean up the gene names
rownames(cc_tumor_fpkm) <- cc_tumor_fpkm$X
current_colname_split <- strsplit(colnames(t(cc_tumor_fpkm)), "_")
finished_gene_list <- c()
for (x in seq(1:length(current_colname_split))) {
  finished_gene_list <- c(finished_gene_list, current_colname_split[[x]][2])
}

cc_tumor_fpkm <- t(cc_tumor_fpkm)
colnames(cc_tumor_fpkm) <- finished_gene_list
cc_tumor_fpkm <- t(cc_tumor_fpkm)
cc_tumor_fpkm <- subset(cc_tumor_fpkm,
  select = c(RHC3546__Tcell__.C6E879:RHC6041__Macrophage__.FFFF55)
)
cc_tumor_fpkm <- apply(cc_tumor_fpkm, c(1, 2), as.numeric)

head(as.data.frame(cc_tumor_fpkm))
```

## Step 5: Pre-processing the scRNA-seq data before sending it to MAGIC. We are only keeping genes expressed in at least 10 cells
```{r magic pre-processing step 1}
keep_rows <- rowSums(cc_tumor_fpkm > 0) > 10
table(keep_rows)
cc_tumor_fpkm <- cc_tumor_fpkm[keep_rows, ]
head(as.data.frame(cc_tumor_fpkm))
```

## Step 5b: Historgram plot of MAGIC pre-processing step 1  
```{r plot of magic pre-processing step 1}
# Looking at the distribution of library sizes
library_size_plot_1 <- ggplot() +
  geom_histogram(aes(x = colSums(cc_tumor_fpkm)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size") +
  scale_y_continuous(limits = c(0, 18), expand = c(0, 0))

library_size_plot_1
```


```{r saving plot of magic pre-processing step 1}
ggsave(
  filename = "~/Documents/Work/PhD Program/Hong Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_1.png",
  device = "png", units = "mm", dpi = 600, width = 20, height = 20,
  plot = library_size_plot_1
)
```

## Step 5c: We normalize the library size with the `library.size.normalize` command from the `phateR` package. This command performs L1 normalization on input data such that the sum of expression values for each cell sums to 1, then returns normalized matrix to the metric space using median UMI count per cell effectively scaling all cells as if they were sampled evenly. We then take the square-root of the dataset to transform the data, but avoid the instabilities and pseudocounts needed when taking the log of the dataset. 
```{r magic pre-processing step 2}
# Normalizing the library size
cc_tumor_fpkm <- library.size.normalize(cc_tumor_fpkm)
head(as.data.frame(cc_tumor_fpkm))
cc_tumor_fpkm <- sqrt(cc_tumor_fpkm)
head(as.data.frame(cc_tumor_fpkm))
```

## Step 5d: Ploting the second MAGIC pre-processing step as a histogram
```{r plot of magic pre-processing step 2}
library_size_plot_2 <- ggplot() +
  geom_histogram(aes(x = colSums(cc_tumor_fpkm)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size Normalized") +
  scale_y_continuous(limits = c(0, 18), expand = c(0, 0))

library_size_plot_2
```

## Step 5e: Saving the plot of the second MAGIC pre-processing step
```{r saving plot of magic pre-processing step 2}
ggsave(
  filename = "~/Documents/Work/PhD Program/Hong Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_2.png",
  device = "png", units = "mm", dpi = 600, width = 20, height = 20,
  plot = library_size_plot_2
)
```

## Step 6a: Denoising the colorectal single-cell data with MAGIC following recommended settings from [here] {http://htmlpreview.github.io/?https://github.com/KrishnaswamyLab/MAGIC/blob/master/Rmagic/inst/examples/bonemarrow_tutorial.html} 
```{r magic denoising of single-cell data}
# Denoising the single-cell data and saving the output
cc_tumor_fpkm <- magic_denoiser(
  sc.data = cc_tumor_fpkm,
  magic.seed = 123, magic.solver = "approximate"
)
```

Step 6b: Saving the outputs of the denoised colorectal single-cell data
```{r saving magic denoising of single-cell data outputs}
saveRDS(cc_tumor_fpkm$denoised_sc_dataframe,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data.rds"
)

head(as.data.frame(cc_tumor_fpkm$denoised_sc_dataframe))

saveRDS(cc_tumor_fpkm$cds_gene_names,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data-gene-names.rds"
)

head(cc_tumor_fpkm$cds_gene_names)
```

## Step 7: Now generating the VIM based pseudotime progression via `Monocle3`. Based on the expresson of VIM we select the root that has the lowest expression of VIM as the starting point of our pseudotime (i.e., where our pseudotime is zero). IF your dataset has a different point with zero expression modify the line `my.root`. 
```{r monocle3 vim pseudotime generation, echo=FALSE}
# Generating the VIM based pseudotime progression with Monocle3
# Based on the expression of the VIM graph we select the bottom-rightmost point
# as the root of our pseudotime as this matches the gradient of VIM expression
# level
cds_output <- cell_dataset_builder(
  vim.genes = c("VIM", "VIMP", "ZEB1", "CDH1", "ANXA4", "ANXA9", "MST1R", "CAPN1", "NQO1", "PLXNB2", "SHANK2", "TSPAN13", "CEACAM1", "CD9", "EPS8L2", "EPHA1", "PTK6", "CLDN4", "EXPH5", "VAMP8", "ATP1B1", "GALNT7"),
  cell.data = my_sc_df,
  cell.meta = my_sc_df_names,
  # cell.data = cc_tumor_fpkm$denoised_sc_dataframe,
  # cell.meta = cc_tumor_fpkm$cds_gene_names,
  my.root = "Y_7",
  point.size = 2.5,
  use.function = FALSE,
  norm.flag = "log",
  my.cds.filename = "Data/Reproducible-results/Rmarkdown/Outputs/cds_ouput_test5.rds",
  my.monocle.graph = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_test5.svg",
  my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/vim_genes_test5.svg",
  my.pt.data.filename = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_data_test5.csv"
)

cds_output$`PT Graph`
cds_output$`Cell Progression Graph`
```

## Step 8: median absolute deviation (`MAD`) metric. 
```{r mad metric}
# MAD metric for colon and rectal cancer
mad.genes <- mad_calculator(cc_tumor_fpkm$denoised_sc_dataframe)
saveRDS(mad.genes,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds"
)

head(mad.genes)
```

## Step 9: switchde (`SDE`) metric. 
```{r sde metric}
sde.genes <- switchde_calculator(cc_tumor_fpkm$denoised_sc_dataframe,
  pseudo.time = cds_output$Pseudotime
)

head(sde.genes)
```


```{r saving sde metric}
saveRDS(sde.genes,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds"
)
```

## Step 10: Downloading the COAD bulk dataset from TCGA. We are using the FPKM (now called STAR - Counts) dataset. See [here]https://docs.gdc.cancer.gov/Data/Release_Notes/Data_Release_Notes/#data-release-320
```{r TCGA-COAD download}
# Now getting our colon cancer bulk data set
# TCGA-COAD
# coad_query <- GDCquery(project       = "TCGA-COAD",
#                        data.category = "Transcriptome Profiling",
#                        data.type     = "Gene Expression Quantification",
#                        workflow.type = "STAR - Counts")
#
# #Downloading the files
# GDCdownload(query           = coad_query,
#             method          = "api",
#             files.per.chunk = 5,
#             directory       = "Outputs/Bulk-data/")
#
#
# #Making the SummarizedExperiment object
# coad_data_se <- GDCprepare(coad_query, summarizedExperiment = TRUE,
#                            directory = "Outputs/Bulk-data/")
# coad_data_df <- as.data.frame(colData(coad_data_se))
# coad_data_df$vital_status <- factor(coad_data_df$vital_status,
#                                     levels = c("Alive", "Dead"),
#                                     labels = c(0,1))
# coad_data_df$vital_status <- as.numeric(as.character(coad_data_df$vital_status))
#
#
# bulk_rna_df <- coad_data_se@assays@data@listData[["STAR - Counts"]]
# colnames(bulk_rna_df) <- coad_data_se@colData@rownames
# rownames(bulk_rna_df) <- coad_data_se@rowRanges@elementMetadata@listData[["external_gene_name"]]
# bulk_rna_df <- t(bulk_rna_df)
# bulk_rna_df <- as.data.frame(bulk_rna_df)
# bulk_rownames <- rownames(bulk_rna_df)
# bulk_rna_df$barcode <- bulk_rownames
#
# bulk_rna_df_unique <- subset(bulk_rna_df,
#                              select = unique(colnames(bulk_rna_df)))
# coad_data_df_unique <- subset(coad_data_df,
#                               select = unique(colnames(coad_data_df)))
# merged_df <- merge(bulk_rna_df_unique, coad_data_df_unique, by = 'barcode')
# rownames(merged_df) <- merged_df$barcode
# merged_df <- merged_df[,2:length(colnames(merged_df))]
#
#
#
# merged_df$days_to_last_follow_up <- ifelse(merged_df$vital_status==1,
#                                            merged_df$days_to_death,
#                                            merged_df$days_to_last_follow_up)
#
# merged_df <- filter(merged_df, days_to_last_follow_up != "NA")
#
#
# cox_time <- merged_df$days_to_last_follow_up
# cox_event <- merged_df$vital_status
# cox_tumor <- merged_df$ajcc_pathologic_stage
# cox_tumor_n <- merged_df$ajcc_pathologic_n
# cox_tumor_m <- merged_df$ajcc_pathologic_m
# cox_gender <- merged_df$gender
# cox_eth <- merged_df$ethnicity
# cox_race <- merged_df$race
# cox_type <- merged_df$definition
# cox_df <- subset(merged_df, select=c(TSPAN6:AC007389.5))
# cox_df$days.to.last.follow.up <- cox_time
# cox_df$vital.status <- cox_event
# cox_df$tumor.stage <- cox_tumor
# cox_df$ajcc.m <- cox_tumor_m
# cox_df$ajcc.n <- cox_tumor_n
# cox_df$race <- cox_race
# cox_df$ethnicity <- cox_eth
# cox_df$gender <- cox_gender
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="A", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="B", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="C", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage IV", replacement = 4)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage III", replacement = 3)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage II", replacement = 2)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage I", replacement = 1)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="a", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="b", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="c", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N0", replacement=0)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N1", replacement=1)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N2", replacement=2)
# cox_df$sample.type <- cox_type
# cox_df <- filter(cox_df, !tumor.stage=="not reported")
# cox_df <- cox_df[complete.cases(cox_df[, "ajcc.m"]), ]
# cox_df$days.to.last.follow.up <- ifelse(cox_df$days.to.last.follow.up < 1, 1,
#                                         cox_df$days.to.last.follow.up)
#
# saveRDS(cox_df, "Data/Reproducible-results/Rmarkdown/Outputs/coad_df.rds")

cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

head(cox_df)
```

## Step 11a: Getting ideal gene number for `MAD` metric on TCGA-COAD
```{r ideal gene number for mad on COAD}
# Getting ideal gene number for MAD metric on TCGA-COAD
gene_sizes <- seq(100, 6000, 50)
mad_cindices <- rep(0, length(gene_sizes))

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = mad.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/mad_coad_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mad_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mad_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mad_cindices_coad_df <- as.data.frame(cbind(gene_sizes, mad_cindices))
mad_cindices_coad_df$method <- rep("MAD", nrow(mad_cindices_coad_df))
colnames(mad_cindices_coad_df)[2] <- "c_index"
```

## Step 11b: Saving the data frame of the concordance index at various gene sizes for `MAD` metric on TCGA-COAD
```{r saving ideal gene number for mad on coad}
write.csv(mad_cindices_coad_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/mad_cindices_coad_across_gene_size.csv"
)
```

## Step 11c: Getting the gene size that maximizes the concordance index of `MAD` on COAD
```{r getting the gene size that maximizes c-index of MAD on COAD}
mad_coad <- read.csv("~/Desktop/mad_cindices_coad_across_gene_size.csv")
mad_coad[which.max(mad_coad$c_index), ]
```


## Step 12a: Getting ideal gene number for `SDE` metric on TCGA-COAD
```{r ideal gene number for sde on coad}
# Getting ideal gene number for SDE metric on TCGA-COAD
gene_sizes <- seq(100, 3000, 50)
sde_cindices <- rep(0, length(gene_sizes))

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = sde.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/sde_coad_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the sde_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  sde_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
sde_cindices_coad_df <- as.data.frame(cbind(gene_sizes, sde_cindices))
sde_cindices_coad_df$method <- rep("SDE", nrow(sde_cindices_coad_df))
colnames(sde_cindices_coad_df)[2] <- "c_index"
```

## Step 12b: Saving the data frame of the concordance index at various gene sizes for `SDE` metric on TCGA-COAD 
```{r saving ideal gene number for sde on coad}
write.csv(sde_cindices_coad_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/sde_cindices_coad_across_gene_size.csv"
)
```

## Step 12c: Getting the gene size that maximizes the concordance index of `SDE` on COAD
```{r getting the gene size that maximizes c-index of SDE on COAD}
sde_coad <- read.csv("~/Desktop/sde_cindices_coad_across_gene_size.csv")
sde_coad[which.max(sde_coad$c_index), ]
```


## Step 13a: Getting the ideal gene number for the high `miRNA` metric on TCGA-COAD. 
For this metric we don't know which combination of miRNA and miRNA targets
will yield the best result in advance so we are trying 3 different miRNA-
miRNA target combinations that encompass specific areas of our grid search.
They are low miRNA number and low miRNA target number, medium miRNA number and
medium miRNA target number, and high miRNA number and high miRNA target number.
Once the ideal miRNA-miRNA target pair is known from the grid search we will
also include a gene size search for it here


```{r high mirna small target file generation}
coad_mirnas <- mirna_calculator(
  save.mirna.genes = TRUE,
  mirna.ranking.name = "~/Desktop/high_mirna.rds",
  mirna.remove = c("hsa-miR-129-2-3p", "hsa-miR-129-1-3p"),
  mirna.genes.mat.name = "~/Desktop/mirna_high_matrix.rds",
  save.mirna.genes.mat = TRUE
)
```
```{r getting overlap of mirnas of tcga-coad and our mirnas}
mirna_high_matrix <- readRDS("~/Desktop/mirna_high_matrix.rds")
miRNA_sums <- colSums(mirna_high_matrix)
miRNA_sums_ranked <- sort(miRNA_sums, decreasing = TRUE)
tcga_coad_mirna <- coad_data_se$miRNA_ID
my_mirna_coad <- names(miRNA_sums_ranked)
my_mirna_coad <- gsub(pattern = "miR", replacement = "mir", x = my_mirna_coad)
mirna_overlap <- intersect(my_mirna_coad, tcga_coad_mirna)
mirna_overlap2 <- intersect(tcga_coad_mirna, my_mirna_coad)
perc_overlap1 <- length(mirna_overlap) / 1881 * 100
perc_overlap2 <- length(mirna_overlap2) / 1881 * 100
perc_overlap3 <- length(mirna_overlap) / 1320 * 100
perc_overlap4 <- length(mirna_overlap2) / 1320 * 100

# Percent overlap ranges from 24.61% when using the number of miRNAS
# from TCGA-COAD dataset to 35.08% when using the the number of miRNAs
# in our metric
perc_overlap1
perc_overlap2
perc_overlap3
perc_overlap4
```


```{r ideal gene number for high miRNA metric on COAD}
# Getting ideal gene number for miRNA metric on TCGA-COAD
# For this metric we don't know which combination of miRNA and miRNA targets
# will yield the best result in advance so we are trying 3 different miRNA-
# miRNA target combinations that encompass specific areas of our grid search.
# They are low miRNA number and low miRNA target number, medium miRNA number and
# medium miRNA target number, and high miRNA number and high miRNA target number.
# Once the ideal miRNA-miRNA target pair is known from the grid search we will
# also include a gene size search for it here
gene_sizes <- seq(100, 3000, 50)
mirna_high_cindices <- rep(0, length(gene_sizes))
mirna_medium_cindices <- rep(0, length(gene_sizes))
mirna_low_cindices <- rep(0, length(gene_sizes))

# Loading the high miRNA-miRNA target file
load("Data/Reproducible-results/Rmarkdown/Data/800_1010_targets.RData",
  verbose = TRUE
)

high.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = high.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/high_mirna_coad_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_high_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_high_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_high_cindices_coad_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_high_cindices
))

colnames(mirna_high_cindices_coad_df)[2] <- "c_index"
mirna_high_cindices_coad_df$mirna_type <- rep(
  "high",
  nrow(mirna_high_cindices_coad_df)
)

write.csv(
  mirna_high_cindices_coad_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_high_cindices_coad_across_gene_size.csv"
)
```

## Step 13b: Getting the ideal gene number for the medium `miRNA` metric on TCGA-COAD
```{r ideal gene number for medium miRNA metric on COAD}
# Medium miRNA-miRNA target number
# Loading the medium miRNA-miRNA target file
load("Data/Reproducible-results/Rmarkdown/Data/400_510_targets.RData", verbose = TRUE)

medium.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = medium.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/medium_mirna_coad_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_medium_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_medium_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_medium_cindices_coad_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_medium_cindices
))

colnames(mirna_medium_cindices_coad_df)[2] <- "c_index"
mirna_medium_cindices_coad_df$mirna_type <- rep(
  "medium",
  nrow(mirna_medium_cindices_coad_df)
)
```


```{r saving ideal gene number for medium miRNA metric on COAD}
write.csv(
  mirna_medium_cindices_coad_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_medium_cindices_coad_across_gene_size.csv"
)
```

## Step 13c: Getting the ideal gene number for the low `miRNA` metric on TCGA-COAD
```{r ideal gene number for low miRNA metric on coad}
# low miRNA-miRNA target number
# Loading the low miRNA-miRNA target file
load("Data/Reproducible-results/Data/100_10_targets.RData", verbose = TRUE)

low.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = low.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/low_mirna_coad_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_medium_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_low_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_low_cindices_coad_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_low_cindices
))

colnames(mirna_low_cindices_coad_df)[2] <- "c_index"
mirna_low_cindices_coad_df$mirna_type <- rep(
  "low",
  nrow(mirna_low_cindices_coad_df)
)
```


```{r saving ideal gene number for low miRNA metric on coad}
write.csv(
  mirna_low_cindices_coad_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_low_cindices_coad_across_gene_size.csv"
)
```

## Step 13d: Making one large data frame with all 3 type of `miRNA` in it 
```{r mirna metric dataframe made}
# Binding all the rows of the 3 different miRNA data frames together into 1 big
# data frame
all_mirna_metrics_coad_df <- bind_rows(
  mirna_low_cindices_coad_df,
  mirna_medium_cindices_coad_df,
  mirna_high_cindices_coad_df
)


# Reordering the labels to make them look nicer on the plot
all_mirna_metrics_coad_df$mirna_type <- factor(all_mirna_metrics_coad_df$mirna_type,
  levels = c("low", "medium", "high")
)
```

## Step 13e: Plotting the three kinds of `miRNA` metric for COAD data
```{r ploting of mirna metric levels on coad}
# Basic miRNA-miRNA target plot
all_mirna_metrics_coad_plot <- ggplot(
  data = all_mirna_metrics_coad_df,
  aes(x = gene_sizes, y = c_index, color = mirna_type)
)

# Making the plot much nicer
all_mirna_metrics_coad_plot_finished <- all_mirna_metrics_coad_plot +
  geom_line(size = 2.5) + geom_point(size = 3.0) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(size = 40, face = "bold", hjust = 0.5),
    axis.title = element_text(size = 40, face = "bold"),
    legend.title = element_text(size = 30, face = "bold"),
    legend.position = "bottom",
    legend.text = element_text(size = 35, face = "plain"),
    axis.text = element_text(size = 25, face = "bold")
  ) +
  xlab("Gene Size") + ylab("Mean Concordance Index") +
  ggtitle("miRNA-miRNA Target Number") +
  labs(color = "miRNA-miRNA Target Number") +
  scale_color_viridis_d()

all_mirna_metrics_coad_plot_finished
```

## Step 13f: Saving the plot of the different kinds of `miRNA` metric on TCGA-COAD 
```{r saving mirna metric type plot coad}
# Saving the finished graph in .svg format
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/mirna_mirna_target_num_across_gene_size.svg",
  plot = print(all_mirna_metrics_coad_plot_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 34, height = 34,
  units = "mm"
)
```

## Step 13g: Getting the top value of miRNA on COAD
```{r getting ideal gene performance for high mirna COAD}

mirna_high_df <- read.csv("~/Desktop/mirna_high_cindices_coad_across_gene_size.csv")

mirna_high_df[which.max(mirna_high_df$c_index), ]
```

## Step 14: Get `MAD + SDE` best performance for figure 2a on TCGA-COAD data. Here we tried several different values of `alpha` originally and did indeed confirm that `glmnet` documentation was correct for our case and the best performance occurs when `alpha` = 1. The best performance occurs at combo 21 but we test all 88 combos here just to be thorough. Concordance index of 0.6607 at X (verify and determine location)
```{r getting ideal mad+sde gene performance COAD}
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

setwd("~/Documents/Work/PhD Program/Hong Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown")

combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010",
  "600_10", "600_110", "600_210", "600_310", "600_410", "600_510", "600_610", "600_710", "600_810", "600_910", "600_1010",
  "500_10", "500_110", "500_210", "500_310", "500_410", "500_510", "500_610", "500_710", "500_810", "500_910", "500_1010",
  "400_10", "400_110", "400_210", "400_310", "400_410", "400_510", "400_610", "400_710", "400_810", "400_910", "400_1010",
  "300_10", "300_110", "300_210", "300_310", "300_410", "300_510", "300_610", "300_710", "300_810", "300_910", "300_1010",
  "200_10", "200_110", "200_210", "200_310", "200_410", "200_510", "200_610", "200_710", "200_810", "200_910", "200_1010",
  "100_10", "100_110", "100_210", "100_310", "100_410", "100_510", "100_610", "100_710", "100_810", "100_910", "100_1010"
)

alpha_value <- c(0.0, 0.5, 1.0)

gene_sizes <- seq(100, 3000, 50)
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    mad_sde_optimized <- two_weight_optimizer(
      first.metric = mad.genes,
      second.metric = sde.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_mad_sde_", data_set, ".rds")
    )



    index <- 1

    for (md in mad_sde_optimized) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          cox.predictors = md,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("mad_sde_coad_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
        index <- index + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
      index <- index + 1
    }

    write.csv(my_cindices, file = paste0("top_cindices_mad_sde_coad_df.csv"))
  }
}
```

## Step 15: Getting ideal concordance index of random genes on COAD. We do 10 reps of random genes and take the average. Its concordance index performance is 0.63321
```{r random genes for COAD}
# Random sample of genes (10 random samples taken) for COAD.
# We will take the average of these 10 samplings. We will select the same number
# of genes as the number of genes that give our miRNA metric the best performance
# (2,500 genes)
random_coad1 <- sample(colnames(cox_df), size = 2500)
random_coad2 <- sample(colnames(cox_df), size = 2500)
random_coad3 <- sample(colnames(cox_df), size = 2500)
random_coad4 <- sample(colnames(cox_df), size = 2500)
random_coad5 <- sample(colnames(cox_df), size = 2500)
random_coad6 <- sample(colnames(cox_df), size = 2500)
random_coad7 <- sample(colnames(cox_df), size = 2500)
random_coad8 <- sample(colnames(cox_df), size = 2500)
random_coad9 <- sample(colnames(cox_df), size = 2500)
random_coad10 <- sample(colnames(cox_df), size = 2500)


# Saving the random results for reproducibility
# write.csv(random_coad1, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes1.csv")
# write.csv(random_coad2, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes2.csv")
# write.csv(random_coad3, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes3.csv")
# write.csv(random_coad4, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes4.csv")
# write.csv(random_coad5, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes5.csv")
# write.csv(random_coad6, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes6.csv")
# write.csv(random_coad7, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes7.csv")
# write.csv(random_coad8, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes8.csv")
# write.csv(random_coad9, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes9.csv")
# write.csv(random_coad10, "Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes10.csv")


random_coad1 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes1.csv")
random_coad2 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes2.csv")
random_coad3 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes3.csv")
random_coad4 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes4.csv")
random_coad5 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes5.csv")
random_coad6 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes6.csv")
random_coad7 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes7.csv")
random_coad8 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes8.csv")
random_coad9 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes9.csv")
random_coad10 <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/random_coad_genes10.csv")

# Putting all the gene vectors together in a list to loop over for cox model
random_coad_gene_list <- list(
  random_coad1, random_coad2, random_coad3,
  random_coad4, random_coad5, random_coad6,
  random_coad7, random_coad8, random_coad9,
  random_coad10
)

all_random_gene_cindices <- seq(1, 10, 1)

for (rg in all_random_gene_cindices) {
  current_genes <- random_coad_gene_list[[rg]]
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = current_genes,
    cox.df = cox_df,
    gene.num = 2500,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = "random_coad_coefs_for_random_genes.csv"
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the all_random_gene_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  all_random_gene_cindices[rg] <- top_cindex
}

write.csv(all_random_gene_cindices,
  file = "coad_random_genes.csv"
)



mean_random <- mean(all_random_gene_cindices)
```

## Step 16a: Getting ideal gene size for CC Singlecell MS on COAD. It occurs at the 49th gene size value (2500 genes). At this value with the 21st combo used (700_910 with an alpha value of 1 and the 10th value of the weighted linear model [mirna_sde_optimized]) we get a concordance index of 0.7445. 
```{r getting ideal cc singlecell ms (miRNA + SDE) gene performance COAD}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

gene_sizes <- seq(100, 3000, 50)
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[21]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_sde_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = sde.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_ms_", a, "_alpha_", data_set, ".rds")
    )







    for (ms in mirna_sde_optimized[10]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          my.dataset = "COAD",
          cox.predictors = ms,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("cc_singlecell_ms_ideal_gene_size_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_ms_coad_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_ms_coad_df_all_gene_size_data.csv"))
  }
}
```

## Step 16b: Now getting the ideal combo at the optimal gene size (2,500 genes) for CC Singlecell MS on COAD. The ideal peformance is at multiple combos (8-11) which corresponds to "800_710" , "800_810",  "800_910", and "800_1010" which keeps in line with our high miRNA-high miRNA target number giving best performance. The concordance index at these combos is 0.7445.  
```{r ideal mirna combo for cc singlecell ms on COAD}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_sde_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = sde.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_ms_", a, "_alpha_", data_set, ".rds")
    )







    for (ms in mirna_sde_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "COAD",
        cox.predictors = ms,
        cox.df = cox_df,
        gene.num = 2500,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_ms_ideal_mirna_combo_coefs.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_ms_coad_df_ideal_mirna_combo.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_ms_coad_df_all_mirna_combo_performance_data.csv"))
}
```

## Step 17a: Getting ideal gene size for CC Singlecell MM on COAD. It occurs at the 52nd gene size value (2,650 genes). At this value with the 11st combo used (800_1010 with an alpha value of 1 and the 10th value of the weighted linear model [mirna_sde_optimized]) we get a concordance index of 0.7334. 
```{r getting ideal cc singlecell mm (miRNA + MAD) gene performance COAD}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

gene_sizes <- seq(100, 3000, 50)
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[11]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized[10]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          my.dataset = "COAD",
          cox.predictors = mm,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("cc_singlecell_mm_ideal_gene_size_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_mm_coad_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_mm_coad_df_all_gene_size_data.csv"))
  }
}
```

## Step 17b: Now getting the ideal combo at the optimal gene size (2,650 genes) for CC Singlecell MM on COAD.It is the 19th combo ("700_710") with a concordance index peformance of 0.7351
```{r ideal mirna combo for cc singlecell mm on COAD}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "COAD",
        cox.predictors = mm,
        cox.df = cox_df,
        gene.num = 2650,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_mm_ideal_mirna_combo_coefs.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_mm_coad_df_ideal_mirna_combo.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_mm_coad_df_all_mirna_combo_performance_data.csv"))
}
```

## Step 18: Finding performance of CC Singlecell MMS (miRNA + SDE + MAD) on COAD at the ideal points for both CC Singlecell MS and MM to save computation time and see if we improve performance. We use the gene size of CC Singlecell MS because it is the method with the best performance up to this point. With the MM ideal mirna combination and the MS gene size the concordance index is 0.7346. With the MS ideal miRNA combination and the MS gene size the concordance index is 0.7491. After the grid search for COAD in fig S3a we found a slightly better point and it has a concordance index of 0.7514.  
```{r cc singlecell mms coad mm ideal mirna combo}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[19]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_sde_optimized <- three_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      third.metric = sde.genes,
      my.filename = paste0("Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mms in mirna_mad_sde_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "COAD",
        cox.predictors = mms,
        cox.df = cox_df,
        gene.num = 2500,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_mms_ideal_mirna_combo_coefs.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_mms_coad_df_ideal_mirna_combo.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_mms_coad_df_all_mirna_combo_performance_data.csv"))
}
```


```{r cc singlecell mms coad ms ideal mirna combo}
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010"
)

mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "coad"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[11]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_sde_optimized <- three_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      third.metric = sde.genes,
      my.filename = paste0("Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mms in mirna_mad_sde_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "COAD",
        cox.predictors = mms,
        cox.df = cox_df,
        gene.num = 2500,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_mms_ideal_mirna_combo_coefs_ms.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_mms_coad_df_ideal_mirna_combo_ms_combo_used.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_mms_coad_df_all_mirna_combo_performance_data_ms_combo_used.csv"))
}
```


## Step 19a: Bring all metric data together for figure 2a for TCGA-COAD
```{r fig2a data}
coad_methods_comp_df <- data.frame(
  Method = c(
    "CCsc MM",
    "CCsc MS",
    "CCsc MMS", "MAD", "SDE",
    "miRNA", "MAD + SDE",
    "Random Genes"
  ),
  c_index = c(
    0.7351, 0.7445, 0.7514,
    0.6197,
    0.6495,
    0.7325,
    0.6607, 0.63321
  )
)

# Factoring the levels to make the plot nicer
coad_methods_comp_df$method <- factor(coad_methods_comp_df$Method,
  levels = c(
    "CCsc MMS",
    "CCsc MS",
    "CCsc MM", "miRNA",
    "MAD", "SDE", "MAD + SDE",
    "Random Genes"
  )
)



# Saving the data frame
write.csv(coad_methods_comp_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/fig2a_data.csv"
)
```

## Step 19b: Fig 2a plot
```{r fig2a plot}
methods_comp_coad_plot <- ggplot(
  data = coad_methods_comp_df,
  aes(x = method, y = c_index, fill = method)
)

# Making the plot much nicer
methods_comp_coad_plot_finished <- methods_comp_coad_plot +
  geom_col() +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    legend.title = element_text(size = 30, face = "bold"),
    legend.position = "none",
    text = element_text(family = "Arial"),
    legend.text = element_text(size = 35, face = "plain"),
    axis.text.x = element_text(
      size = 22, face = "bold", angle = 90,
    ),
    axis.text.y = element_text(size = 22, face = "bold")
  ) +
  xlab("Method") + ylab("Mean C-index") +
  ggtitle("TCGA-COAD")

methods_comp_coad_plot_finished <- methods_comp_coad_plot_finished + coord_cartesian(ylim = c(0.5, 0.76)) +
  scale_fill_manual(values = c("#FDE725FF", "#FDE725FF", "#FDE725FF", "#404788FF", "#404788FF", "#404788FF", "#404788FF", "#404788FF"))


methods_comp_coad_plot_finished
```


```{r saving fig2a plot}
# Saving the finished graph in .svg format
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig2a.svg",
  plot = print(methods_comp_coad_plot_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 250, height = 250,
  units = "mm"
)
```

## Step 20a: Fig 2c
```{r fig2c (cc singlecell mms coad km curve)}
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")
fig2c_plot <- risk_score_calculator(
  my.file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_ideal_mirna_fig_s3_coefs.csv",
  tumor.data = FALSE, n.data = FALSE,
  set.ci = TRUE,
  cox.df = cox_df,
  plot.title = "TCGA-COAD"
)
fig2c_plot$`KM Plot`
```

## Step 20b: Saving fig 2c
```{r saving fig2c (cc singlecell mms coad km curve)}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig2c.svg",
  plot = print(fig2c_plot$`KM Plot`$plot, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 340, height = 340,
  units = "mm"
)
```


## Step 21a: Fig 4a data
```{r fig4a data (cc singlecell mms COAD coef plot)}
coad_coef_df <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_ideal_mirna_fig_s3_coefs.csv")
coad_coef_df <- coad_coef_df[, 2:3]
colnames(coad_coef_df) <- c("Gene", "coefs")
coad_coef_df$hazard_ratio <- exp(coad_coef_df$coefs)

# Changing the threshold to be > 2 for risk increasing and
# < 0.5 for risk decreasing
coad_coef_df_sub_neg <- filter(coad_coef_df, hazard_ratio > 2)
coad_coef_df_sub_pos <- filter(coad_coef_df, hazard_ratio < 0.5)
coad_coef_df_sub_neg$log_hazard_ratio <- log2(coad_coef_df_sub_neg$hazard_ratio)
coad_coef_df_sub_pos$log_hazard_ratio <- log2(coad_coef_df_sub_pos$hazard_ratio)
```

## Step 21b: Fig 4a plot
```{r fig4a (cc singlecell mms COAD coef plot)}
level_order <- order(coad_coef_df_sub_neg$log_hazard_ratio, decreasing = TRUE)
coad_coef_df_sub_neg_ordered <- coad_coef_df_sub_neg[level_order, ]
coad_coef_plot <- ggplot(data = coad_coef_df_sub_neg, aes(
  x = factor(Gene, levels = rev(c("ZNF705D", "UNC5D", "TP53TG3D", "DNAJC5G", "FABP7", "FAM182A", "KCNC1", "RSPH10B", "HS6ST3", "ST6GALNAC3", "RFPL3S", "ABCA13", "OPCML"))), y = log_hazard_ratio, color = "red",
  fill = "red"
)) +
  geom_col(fill = "red", color = "red") +
  theme_bw() +
  ggtitle("Top Increasing") +
  ylab(bquote("log"[2] ~ "(Hazard Ratio)")) +
  xlab(NULL) +
  theme(
    text = element_text(family = "Arial"),
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5, face = "bold", size = 32
    ),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  coord_flip() +
  scale_y_break(c(9, 150), expand = expansion(mult = c(0, 0.1)), scales = 0.60)


coad_coef_plot
```

## Step 21c: Saving fig 4a plot
```{r saving fig 4a (cc singlecell mms COAD coef plot)}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig4a.svg",
  plot = print(coad_coef_plot),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```
```{r fig4a pos (cc singlecell mms COAD coef plot)}
level_order <- order(coad_coef_df_sub_pos$log_hazard_ratio, decreasing = FALSE)
coad_coef_df_sub_pos_ordered <- coad_coef_df_sub_pos[level_order, ]
coad_coef_plot_pos <- ggplot(data = coad_coef_df_sub_pos, aes(
  x = factor(Gene, levels = rev(c("BSND", "GRAMD4P3", "AJAP1", "SLC24A5", "C1orf229", "HAP1", "AC079612.1", "TENM2", "AL133373.1", "RS1", "CACNG8", "ICOS"))), y = log_hazard_ratio, fill = "pale green",
  color = "pale green"
)) +
  geom_col(fill = "pale green", color = "pale green") +
  theme_bw() +
  ggtitle("Top Decreasing") +
  ylab(bquote("log"[2] ~ "(Hazard Ratio)")) +
  xlab("Genes") +
  theme(
    text = element_text(family = "Arial"),
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5, face = "bold", size = 32
    ),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
  ) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0)), labels = c(-8, -6, -4, -2, 0)) +
  coord_flip()


coad_coef_plot_pos
```

```{r saving fig 4a pos coefs}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig4a_pos.svg",
  plot = print(coad_coef_plot_pos, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```

## Step 22: Downloading the READ bulk dataset
```{r downloading bulk rectal cancer from TCGA}
# Now doing all of these steps for the rectal cancer data set from TCGA-READ
# Now getting our rectal cancer bulk data set
# TCGA-READ
# read_query <- GDCquery(project       = "TCGA-READ",
#                        data.category = "Transcriptome Profiling",
#                        data.type     = "Gene Expression Quantification",
#                        workflow.type = "HTSeq - FPKM")
#
# #Downloading the files
# GDCdownload(query           = read_query,
#             method          = "api",
#             files.per.chunk = 10,
#             directory       = "Data/Reproducible-results/Data/Bulk-data/")
#
#
# #Making the SummarizedExperiment object
# read_data_se <- GDCprepare(read_query, summarizedExperiment = TRUE,
#                            directory = "Data/Reproducible-results/Data/Bulk-data/")
# read_data_df <- as.data.frame(colData(read_data_se))
# read_data_df$vital_status <- factor(read_data_df$vital_status,
#                                     levels = c("Alive", "Dead"),
#                                     labels = c(0,1))
# read_data_df$vital_status <- as.numeric(as.character(read_data_df$vital_status))
#
#
# bulk_rna_df <- read_data_se@assays@data@listData[["HTSeq - FPKM"]]
# colnames(bulk_rna_df) <- read_data_se@colData@rownames
# rownames(bulk_rna_df) <- read_data_se@rowRanges@elementMetadata@listData[["external_gene_name"]]
# bulk_rna_df <- t(bulk_rna_df)
# bulk_rna_df <- as.data.frame(bulk_rna_df)
# bulk_rownames <- rownames(bulk_rna_df)
# bulk_rna_df$barcode <- bulk_rownames
#
# bulk_rna_df_unique <- subset(bulk_rna_df,
#                              select = unique(colnames(bulk_rna_df)))
# read_data_df_unique <- subset(read_data_df,
#                               select = unique(colnames(read_data_df)))
# merged_df <- merge(bulk_rna_df_unique, read_data_df_unique, by = 'barcode')
# rownames(merged_df) <- merged_df$barcode
# merged_df <- merged_df[,2:length(colnames(merged_df))]
#
#
#
# merged_df$days_to_last_follow_up <- ifelse(merged_df$vital_status==1,
#                                            merged_df$days_to_death,
#                                            merged_df$days_to_last_follow_up)
#
# merged_df <- filter(merged_df, days_to_last_follow_up != "NA")
#
#
# cox_time <- merged_df$days_to_last_follow_up
# cox_event <- merged_df$vital_status
# cox_tumor <- merged_df$ajcc_pathologic_stage
# cox_tumor_n <- merged_df$ajcc_pathologic_n
# cox_tumor_m <- merged_df$ajcc_pathologic_m
# cox_gender <- merged_df$gender
# cox_eth <- merged_df$ethnicity
# cox_race <- merged_df$race
# cox_type <- merged_df$definition
# cox_df <- subset(merged_df, select=c(TSPAN6:AC007389.5))
# cox_df$days.to.last.follow.up <- cox_time
# cox_df$vital.status <- cox_event
# cox_df$tumor.stage <- cox_tumor
# cox_df$ajcc.m <- cox_tumor_m
# cox_df$ajcc.n <- cox_tumor_n
# cox_df$race <- cox_race
# cox_df$ethnicity <- cox_eth
# cox_df$gender <- cox_gender
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="A", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="B", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern="C", replacement="")
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage IV", replacement = 4)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage III", replacement = 3)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage II", replacement = 2)
# cox_df$tumor.stage <- gsub(cox_df$tumor.stage, pattern = "Stage I", replacement = 1)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="a", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="b", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="c", replacement="")
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N0", replacement=0)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N1", replacement=1)
# cox_df$ajcc.n <- gsub(cox_df$ajcc.n, pattern="N2", replacement=2)
# cox_df$sample.type <- cox_type
# cox_df <- filter(cox_df, !tumor.stage=="not reported")
# cox_df <- cox_df[complete.cases(cox_df[, "ajcc.m"]), ]
# cox_df$days.to.last.follow.up <- ifelse(cox_df$days.to.last.follow.up < 1, 1,
#                                         cox_df$days.to.last.follow.up)
# saveRDS(cox_df, "~/Documents/PhD Program/Hong Lab/Projects/CC_Singlecell/Data/Reproducible-results/Rmarkdown/read_df_finished.rds")

cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")
```

```{r tcga-read-miRNA download}
read_query_mirna <- GDCquery(
  project = "TCGA-READ",
  workflow.type = "BCGSC miRNA Profiling",
  data.category = "Transcriptome Profiling",
  experimental.strategy = "miRNA-Seq",
  data.type = "miRNA Expression Quantification"
)

# Downloading the files
GDCdownload(
  query = read_query_mirna,
  method = "api",
  files.per.chunk = 5,
  directory = "Outputs/Bulk-data/read-miRNA-data"
)

# Making the dataframe of miRNAS for READ
read_mirna_data <- GDCprepare(read_query_mirna,
  summarizedExperiment = TRUE,
  directory = "Outputs/Bulk-data/read-miRNA-data/"
)
```

```{r high mirna small target file generation for read}
read_mirnas <- mirna_calculator(
  save.mirna.genes = TRUE,
  mirna.ranking.name = "~/Desktop/high_mirna_read.rds",
  mirna.remove = c("hsa-miR-129-2-3p", "hsa-miR-129-1-3p"),
  mirna.genes.mat.name = "~/Desktop/mirna_high_matrix_read.rds",
  save.mirna.genes.mat = TRUE
)
```


```{r read mirna percentage}
mirna_high_matrix <- readRDS("~/Desktop/mirna_high_matrix.rds")
miRNA_sums <- colSums(mirna_high_matrix)
miRNA_sums_ranked <- sort(miRNA_sums, decreasing = TRUE)
tcga_read_mirna <- read_mirna_data$miRNA_ID
my_mirna_read <- names(miRNA_sums_ranked)
my_mirna_read <- gsub(pattern = "miR", replacement = "mir", x = my_mirna_read)
my_mirna_read <- gsub(pattern = "hsa-", replacement = "", x = my_mirna_read)
my_mirna_read <- gsub(pattern = "-5p", replacement = "", x = my_mirna_read)
my_mirna_read <- gsub(pattern = "-3p", replacement = "", x = my_mirna_read)
my_mirna_read <- gsub(pattern = "[^mir]", replacement = "", x = my_mirna_read)

tcga_read_mirna <- gsub(pattern = "hsa-", replacement = "", x = tcga_read_mirna)


mirna_overlap <- intersect(my_mirna_read, tcga_read_mirna)
mirna_overlap2 <- intersect(tcga_read_mirna, my_mirna_read)
perc_overlap1 <- length(mirna_overlap) / 1881 * 100
perc_overlap2 <- length(mirna_overlap2) / 1881 * 100
perc_overlap3 <- length(mirna_overlap) / 1320 * 100
perc_overlap4 <- length(mirna_overlap2) / 1320 * 100

# Percent overlap ranges from 24.61% when using the number of miRNAS
# from TCGA-READ dataset to 35.08% when using the the number of miRNAs
# in our metric
perc_overlap1
perc_overlap2
perc_overlap3
perc_overlap4
```


```{r coad mirnas to targetscan}
# Sending the miRNAs from TCGA-COAD dataset to get the top ten targets to compare the targets to the targets of our miRNAs. We are re-purposing a lot of the code from our mirna_calculator_original.R code for this

common_mirnas <- tcga_coad_mirna[12:1881]
common_mirnas <- common_mirnas[!common_mirnas %in% c("hsa-mir-1271", "hsa-mir-129-1")]
# common_mirnas <- gsub(pattern = "hsa-", replacement = "", x = common_mirnas)

my_num <- 1
miRNA_targets <- vector(mode = "list", length = length(common_mirnas))
for (m in common_mirnas[1:length(common_mirnas)]) {
  percent_done <- (my_num / length(common_mirnas)) * 100
  print(percent_done)
  print(m)
  current_target <- targetScan(
    mirna = common_mirnas[my_num],
    species = "Human",
    release = 7.2,
    maxOut = 10
  )
  miRNA_name <- m
  miRNA_name_final <- rep(miRNA_name, times = length(current_target$Ortholog))
  current_target <- cbind(current_target, miRNA_name_final)
  miRNA_targets[[m]] <- current_target
  my_num <- my_num + 1
}



# Simplifying the output of the TargetScan commands to
# just the Gene name and the miRNA columns
counter <- 1
total_list <- list()
for (i in miRNA_targets) {
  current_df <- i
  gene_list <- current_df$Ortholog
  mirna_list <- current_df$miRNA_name_final
  simple_list <- cbind(gene_list, mirna_list)
  total_list[[counter]] <- simple_list
  counter <- counter + 1
}

counter <- 1
all_miRs_for_score <- list()
all_genes_for_score <- list()
for (l in total_list) {
  current_df <- l
  current_miRs <- current_df[, "mirna_list"]
  all_miRs_for_score[[counter]] <- unique(current_miRs)
  current_genes <- current_df[, "gene_list"]
  all_genes_for_score[[counter]] <- current_genes
  counter <- counter + 1
}

all_miRs_for_score <- unlist(all_miRs_for_score)
all_genes_for_score <- unlist(all_genes_for_score)
all_genes_for_score_unique <- unique(all_genes_for_score)

miRNA_score <- matrix(
  data = 0, nrow = length(all_genes_for_score_unique),
  ncol = length(all_miRs_for_score),
  dimnames = list(
    all_genes_for_score_unique,
    all_miRs_for_score
  )
)
miRNA_score <- as.data.frame(miRNA_score)

# Checking to see for each miRNA (column name) if it interacts with a particular row.
# If it does it get a plus one to that cell. If it does not it moves to next cell
for (i in rownames(miRNA_score)) {
  for (x in miRNA_targets) {
    current_df <- x
    if (i %in% current_df$Ortholog) {
      miRNA_to_add <- unique(current_df$miRNA_name_final)
      miRNA_score[i, miRNA_to_add] <- miRNA_score[i, miRNA_to_add] + 1
    }
  }
}

if (save.mirna.genes.mat == TRUE) {
  saveRDS(miRNA_score, file = mirna.genes.mat.name)
}

# Now calculating the row sums of each gene for total number of
# the miRNA interactions----
mirna_gene_list <- rowSums(miRNA_score)
mirna.ranking <- abs(mirna_gene_list) / sum(abs(mirna_gene_list))
mirna.ranking <- sort(mirna.ranking, decreasing = TRUE)
if (save.mirna.genes == TRUE) {
  saveRDS(mirna.ranking, file = mirna.ranking.name)
}
```


## Step 23: Ideal MAD gene size on READ. The concordance index is 0.7543 at 200 genes
```{r mad ideal gene size on READ}
# Getting ideal gene number for MAD metric on TCGA-READ
gene_sizes <- seq(100, 3000, 50)
mad_cindices <- rep(0, length(gene_sizes))

# Loading in the MAD file if they aren't already loaded into the
# environment from earlier
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Data/mad_colon_and_rectal_cancer.rds")


for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = mad.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/mad_read_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mad_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mad_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mad_cindices_read_df <- as.data.frame(cbind(gene_sizes, mad_cindices))
mad_cindices_read_df$method <- rep("MAD", nrow(mad_cindices_read_df))
colnames(mad_cindices_read_df)[2] <- "c_index"
write.csv(mad_cindices_read_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/mad_cindices_read_across_gene_size.csv"
)
```

## Step 24: Ideal gene size for SDE on READ. Its concordance index is 0.7716 at 1,050 genes 
```{r sde ideal gene size on TCGA-READ}
# Getting ideal gene number for SDE metric on TCGA-READ
gene_sizes <- seq(100, 3000, 50)
sde_cindices <- rep(0, length(gene_sizes))

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = sde.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/sde_read_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the sde_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  sde_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
sde_cindices_read_df <- as.data.frame(cbind(gene_sizes, sde_cindices))
sde_cindices_read_df$method <- rep("SDE", nrow(sde_cindices_read_df))
colnames(sde_cindices_read_df)[2] <- "c_index"
write.csv(sde_cindices_read_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/sde_cindices_read_across_gene_size.csv"
)
```


## Step 25a: High-high miRNA on READ. Concordance index is 0.7512 at 2,750 genes when using (1000_1010), 0.7898 at 350 genes when using (1000_110), and 0.7234 at 2,400 genes when using (800_1010)
```{r ideal gene number for high miRNA-miRNA target size on TCGA-READ}
# Getting ideal gene number for miRNA metric on TCGA-READ
# For this metric we don't know which combination of miRNA and miRNA targets
# will yield the best result in advance so we are trying 3 different miRNA-
# miRNA target combinations that encompass specific areas of our grid search.
# They are low miRNA number and low miRNA target number, medium miRNA number and
# medium miRNA target number, and high miRNA number and high miRNA target number.
# Once the ideal miRNA-miRNA target pair is known from the grid search we will
# also include a gene size search for it here. The top performer is 0.7512
gene_sizes <- seq(100, 3000, 50)
mirna_high_cindices <- rep(0, length(gene_sizes))
mirna_medium_cindices <- rep(0, length(gene_sizes))
mirna_low_cindices <- rep(0, length(gene_sizes))

# Loading the high miRNA-miRNA target file
load("Data/Reproducible-results/Rmarkdown/Data/800_1010_targets.RData",
  verbose = TRUE
)
load("Data/Reproducible-results/Data/1000_110_targets.RData", verbose = TRUE)
load("Data/Reproducible-results/Data/1000_1010_targets.RData", verbose = TRUE)

high.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = high.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/high_mirna_read_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_high_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_high_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_high_cindices_read_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_high_cindices
))

colnames(mirna_high_cindices_read_df)[2] <- "c_index"
mirna_high_cindices_read_df$mirna_type <- rep(
  "high",
  nrow(mirna_high_cindices_read_df)
)

write.csv(
  mirna_high_cindices_read_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_high_cindices_read_across_gene_size_test4.csv"
)
```

## Step 25b: Medium-medium miRNA on READ
```{r ideal gene number for medium miRNA-miRNA target number on TCGA-READ}
# Medium miRNA-miRNA target number
# Loading the medium miRNA-miRNA target file
load("Data/Reproducible-results/Rmarkdown/Data/400_510_targets.RData",
  verbose = TRUE
)

medium.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = medium.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/medium_mirna_read_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_medium_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_medium_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_medium_cindices_read_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_medium_cindices
))

colnames(mirna_medium_cindices_read_df)[2] <- "c_index"
mirna_medium_cindices_read_df$mirna_type <- rep(
  "medium",
  nrow(mirna_medium_cindices_read_df)
)

write.csv(
  mirna_medium_cindices_read_df,
  "Data/Reproducible-results/Rmarkdown/mirna_medium_cindices_read_across_gene_size.csv"
)
```

## Step 25c: Low-low miRNA on READ
```{r ideal gene number for low miRNA-miRNA target number on TCGA-READ}
# low miRNA-miRNA target number
# Loading the low miRNA-miRNA target file
load("Data/Reproducible-results/Rmarkdown/Data/100_10_targets.RData", verbose = TRUE)

low.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = low.mirna.genes,
    cox.df = cox_df,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/low_mirna_read_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_medium_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_low_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_low_cindices_read_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_low_cindices
))

colnames(mirna_low_cindices_read_df)[2] <- "c_index"
mirna_low_cindices_read_df$mirna_type <- rep(
  "low",
  nrow(mirna_low_cindices_read_df)
)

write.csv(
  mirna_low_cindices_read_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_low_cindices_read_across_gene_size.csv"
)
```


```{r ideal miRNA data on TCGA-READ}
# Binding all the rows of the 3 different miRNA data frames together into 1 big
# data frame
all_mirna_metrics_read_df <- bind_rows(
  mirna_low_cindices_read_df,
  mirna_medium_cindices_read_df,
  mirna_high_cindices_read_df
)


# Reordering the labels to make them look nicer on the plot
all_mirna_metrics_read_df$mirna_type <- factor(all_mirna_metrics_read_df$mirna_type,
  levels = c("low", "medium", "high")
)
```


```{r ideal miRNA plot on TCGA-READ}
# Basic miRNA-miRNA target plot
all_mirna_metrics_read_plot <- ggplot(
  data = all_mirna_metrics_read_df,
  aes(
    x = gene_sizes, y = c_index,
    color = mirna_type
  )
)

# Making the plot much nicer
all_mirna_metrics_read_plot_finished <- all_mirna_metrics_read_plot +
  geom_line(size = 2.5) + geom_point(size = 3.0) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(size = 40, face = "bold", hjust = 0.5),
    axis.title = element_text(size = 40, face = "bold"),
    legend.title = element_text(size = 30, face = "bold"),
    legend.position = "bottom",
    legend.text = element_text(size = 35, face = "plain"),
    axis.text = element_text(size = 25, face = "bold")
  ) +
  xlab("Gene Size") + ylab("Mean Concordance Index") +
  ggtitle("miRNA-miRNA Target Number") +
  labs(color = "miRNA-miRNA Target Number") +
  scale_color_viridis_d()
```


```{r saving ideal miRNA plot on TCGA-READ}
# Saving the finished graph in .svg format
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/mirna_mirna_target_num_across_gene_size_read.svg",
  plot = print(all_mirna_metrics_read_plot_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 34, height = 34,
  units = "mm"
)
```

## Step 26: Random gene performance on READ. Its c-index is 0.64411
```{r random gene performance for TCGA-READ}
# Random sample of genes (10 random samples taken) for READ.
# We will take the average of these 10 samplings. We will select the same number
# of genes as the number of genes that give our miRNA metric the best performance
# (2,750 genes)
random_read1 <- sample(colnames(cox_df), size = 600)
random_read2 <- sample(colnames(cox_df), size = 600)
random_read3 <- sample(colnames(cox_df), size = 600)
random_read4 <- sample(colnames(cox_df), size = 600)
random_read5 <- sample(colnames(cox_df), size = 600)
random_read6 <- sample(colnames(cox_df), size = 600)
random_read7 <- sample(colnames(cox_df), size = 600)
random_read8 <- sample(colnames(cox_df), size = 600)
random_read9 <- sample(colnames(cox_df), size = 600)
random_read10 <- sample(colnames(cox_df), size = 600)

# Saving the random results for reproducibility
# write.csv(random_read1, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes1.csv")
# write.csv(random_read2, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes2.csv")
# write.csv(random_read3, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes3.csv")
# write.csv(random_read4, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes4.csv")
# write.csv(random_read5, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes5.csv")
# write.csv(random_read6, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes6.csv")
# write.csv(random_read7, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes7.csv")
# write.csv(random_read8, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes8.csv")
# write.csv(random_read9, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes9.csv")
# write.csv(random_read10, "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes10.csv")

# Putting all the gene vectors together in a list to loop over for cox model
random_read_gene_list <- list(
  random_read1, random_read2, random_read3,
  random_read4, random_read5, random_read6,
  random_read7, random_read8, random_read9,
  random_read10
)

# saveRDS(random_read_gene_list, file = "Data/Reproducible-results/Rmarkdown/Outputs/random_read_genes.rds")
```


```{r ideal random gene performance for TCGA-READ}
all_random_gene_cindices <- seq(1, 10, 1)

for (rg in all_random_gene_cindices) {
  current_genes <- random_read_gene_list[[rg]]
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = current_genes,
    cox.df = cox_df,
    gene.num = 600,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/random_read_coefs_for_random_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the all_random_gene_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  all_random_gene_cindices[rg] <- top_cindex
}

write.csv(
  all_random_gene_cindices,
  "Data/Reproducible-results/Rmarkdown/Outputs/read_random_genes.csv"
)

mean_random <- mean(all_random_gene_cindices)
```

## Step 27: Ideal MAD + SDE performance on READ. It's c-index is 0.8107 at 228 
```{r getting ideal MAD + SDE gene performance READ}
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

setwd("~/Documents/Work/PhD Program/Hong Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown")

combo_used <- "1000_1010"

alpha_value <- c(0.0, 0.5, 1.0)

gene_sizes <- seq(100, 3000, 50)
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[11]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    mad_sde_optimized <- two_weight_optimizer(
      first.metric = mad.genes,
      second.metric = sde.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_mad_sde", data_set, ".rds")
    )



    index <- 1

    for (md in mad_sde_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          cox.predictors = md,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("mad_sde_read_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
        index <- index + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
      index <- index + 1
    }

    write.csv(my_cindices, file = paste0("mad_sde_read_df.csv"))
  }
}
```


## Step 28a: Ideal CC Singlecell MM gene size on READ. Its ideal weighted index is 11 (0.7898) and occurs at a gene size of 350 genes
```{r ideal cc singlecell mm gene size READ}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010"
)

gene_sizes <- seq(100, 3000, 50)
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[2]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          my.dataset = "READ",
          cox.predictors = mm,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("cc_singlecell_mm_ideal_gene_size_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_mm_read_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_mm_read_df_all_gene_size_data.csv"))
  }
}
```

## Step 28b: Finding the best miRNA-miRNA target combo at the ideal gene size determined in 28a for CC Singlecell MM on READ. Its ideal c-index is 0.802 and occurs at the "1000_610" miRNA-miRNA target combo
```{r ideal mirna combo for cc singlecell mm on READ}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010"
)


mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "READ",
        cox.predictors = mm,
        cox.df = cox_df,
        gene.num = 350,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_mm_ideal_mirna_combo_coefs.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_mm_read_df_ideal_mirna_combo.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_mm_read_df_all_mirna_combo_performance_data.csv"))
}
```


## Step 29a: Ideal CC Singlecell MS gene size on READ. Its ideal weighted index is 11 (0.7898) and occurs at a gene size of 350 genes 
```{r ideal cc singlecell mm gene size READ}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010"
)

gene_sizes <- seq(100, 3000, 50)
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[2]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_ms_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          my.dataset = "READ",
          cox.predictors = mm,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          my.filename = paste0("cc_singlecell_ms_ideal_gene_size_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_ms_read_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_ms_read_df_all_gene_size_data.csv"))
  }
}
```

##Step 29b: Finding the best miRNA-miRNA target combo at the ideal gene size determined in 29a for CC Singlecell MS on READ. Its ideal c-index is 0.802 and occurs at the "1000_610" miRNA-miRNA target combo. A higher ideal (0.8283) is found at 600 genes and an miRNA-miRNA target value of "400_410". Need to investigate this more 
```{r ideal mirna combo for cc singlecell ms on READ}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010"
)


sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimization_", c, "_targets_cc_singlecell_ms_", a, "_alpha_", data_set, ".rds")
    )







    for (ms in mirna_sde_optimized) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "READ",
        cox.predictors = ms,
        cox.df = cox_df,
        gene.num = 350,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_ms_ideal_mirna_combo_coefs.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }

  write.csv(top_cindices, file = paste0("cc_singlecell_ms_read_df_ideal_mirna_combo.csv"))
  write.csv(my_cindices, file = paste0("cc_singlecell_ms_read_df_all_mirna_combo_performance_data.csv"))
}
```


## Step 30: CC Singlecell MMS on READ: Its best performance is 0.8442. It occurs at gene size of 600. 
```{r ideal gene size cc singlecell mms on READ}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010",
  "600_10", "600_110", "600_210", "600_310", "600_410", "600_510", "600_610", "600_710", "600_810", "600_910", "600_1010",
  "500_10", "500_110", "500_210", "500_310", "500_410", "500_510", "500_610", "500_710", "500_810", "500_910", "500_1010",
  "400_10", "400_110", "400_210", "400_310", "400_410", "400_510", "400_610", "400_710", "400_810", "400_910", "400_1010",
  "300_10", "300_110", "300_210", "300_310", "300_410", "300_510", "300_610", "300_710", "300_810", "300_910", "300_1010",
  "200_10", "200_110", "200_210", "200_310", "200_410", "200_510", "200_610", "200_710", "200_810", "200_910", "200_1010",
  "100_10", "100_110", "100_210", "100_310", "100_410", "100_510", "100_610", "100_710", "100_810", "100_910", "100_1010"
)


alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "read"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used) {
    my_cindices <- rep(0, 121)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-READ/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_sde_optimized <- three_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      third.metric = sde.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_mms_", a, "_alpha_", data_set, ".rds")
    )







    for (mms in mirna_mad_sde_optimized[1:121]) {
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = a,
        my.dataset = "READ",
        cox.predictors = mms,
        cox.df = cox_df,
        gene.num = 600,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        my.filename = paste0("cc_singlecell_mms_", data_set, "_coefs_genes_", mms, "_index.csv")
      )

      # Getting the top concordance index from the cross validation and then rounding
      # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
      # the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
      my_cindices[counter] <- current_cindex
      counter <- counter + 1
    }

    top_cindex <- max(my_cindices)
    top_index <- which(my_cindices == top_cindex)
    print(top_index)
    print(top_cindex)
    top_index_used <- top_index[1]
    top_cindices <- c(top_cindices, top_cindex)
  }


  write.csv(top_cindices, file = paste0("cc_singlecell_mms_read_top_cindex.csv"))
}
```


## Step 31a: Figure 2b data
```{r fig2b data}
# Data for MAD, SDE, and miRNA high metric, CC Singlecell MS, CC Singlecell MM,
# CC Singlecell MMS, MAD + SDE, and Random genes
# First bind the data frames together
methods_optimal_gene_read_df <- data.frame(
  method = c(
    "CCsc MM",
    "CCsc MS",
    "CCsc MMS",
    "MAD", "SDE", "miRNA",
    "MAD + SDE",
    "Random Genes"
  ),
  c_index = c(
    0.802, 0.8283, 0.8442,
    0.7543, 0.7716, 0.7512,
    0.8107,
    0.58279
  )
)


# Saving the data frame
write.csv(methods_optimal_gene_read_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/methods_read_cindex.csv"
)
```

## Step 31b: Figure 2b plot
```{r fig2b plot}
# Basic plot
methods_optimal_gene_read_df$method <- factor(methods_optimal_gene_read_df$method,
  levels = c(
    "CCsc MMS",
    "CCsc MS",
    "CCsc MM",
    "miRNA", "MAD", "SDE",
    "MAD + SDE",
    "Random Genes"
  )
)

methods_optimal_gene_read_plot <- ggplot(
  data = methods_optimal_gene_read_df,
  aes(x = method, y = c_index, fill = method)
)

# Making the plot much nicer
methods_optimal_gene_read_plot_finished <- methods_optimal_gene_read_plot +
  geom_col() +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    legend.title = element_text(size = 24, face = "bold"),
    legend.position = "none",
    text = element_text(family = "Arial"),
    legend.text = element_text(size = 22, face = "plain"),
    axis.text.x = element_text(size = 22, face = "bold", angle = 90),
    axis.text.y = element_text(size = 22, face = "bold")
  ) +
  xlab("Method") + ylab("Mean C-index") +
  ggtitle("TCGA-READ")

methods_optimal_gene_read_plot_finished <- methods_optimal_gene_read_plot_finished + coord_cartesian(ylim = c(0.5, 0.85)) +
  scale_fill_manual(values = c("#FDE725FF", "#FDE725FF", "#FDE725FF", "#404788FF", "#404788FF", "#404788FF", "#404788FF", "#404788FF"))
```

## Step 31c: Saving figure 2b plot
```{r saving fig2b}
# Saving the finished graph in .svg format
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig2b.svg",
  plot = print(methods_optimal_gene_read_plot_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 250, height = 250,
  units = "mm"
)
```

## Step 32a: Figure 2d
```{r cc singlecell mms read km plot (fig2d)}
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")

fig2d_plot <- risk_score_calculator(
  my.file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_read_coefs_at_top_combo_and_index.csv",
  tumor.data = FALSE, n.data = FALSE,
  set.ci = TRUE,
  cox.df = cox_df,
  plot.title = "TCGA-READ"
)
fig2d_plot$`KM Plot`
```

## Step 32b: Saving figure 2d
```{r saving fig2d (cc singlecell mms read km curve)}
# Saving the KM plot
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig2d.svg",
  plot = print(fig2d_plot$`KM Plot`$plot, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 340, height = 340,
  units = "mm"
)
```
## Step 33a: Figure 3a data preprocessing steps for FPKM varient of the file
Now making a combined data frame that is used by many of the other methods to get a concordance index. Some use the FPKM variety and some, e.g. DESeq2 and scDD use the same file but of the COUNT variety
```{r other methods fpkm data prep}
# Loading needed functions for tidying the gene names
gene_name_cleaner <- function(data.to.clean = all_tumor_cells_fpkm_denoised_df) {
  data.to.clean <- t(data.to.clean)
  current_colname_split <- strsplit(colnames(data.to.clean), "_")
  finished_gene_list <- c()
  current_list <- current_colname_split
  for (y in seq(1:length(current_list))) {
    finished_gene_list <- c(finished_gene_list, current_list[[y]][2])
  }
  colnames(data.to.clean) <- finished_gene_list
  return(data.to.clean)
}

# Method to clean just a vector of gene names
gene_vector_cleaner <- function(unclean.data = deseq2_rows) {
  split_data <- strsplit(unclean.data, "_")
  finished_genes <- c()
  current_list <- split_data
  for (y in seq(1:length(current_list))) {
    finished_genes <- c(finished_genes, current_list[[y]][2])
  }
  unclean.data <- finished_genes
  return(unclean.data)
}


all_tumor_cells_fpkm <- read.csv("Data/Single-cell-data/FPKM/GSE81861_CRC_tumor_all_cells_FPKM.csv")
rownames(all_tumor_cells_fpkm) <- all_tumor_cells_fpkm$X
all_tumor_cells_fpkm <- gene_name_cleaner(data.to.clean = all_tumor_cells_fpkm)
all_tumor_cells_fpkm <- t(all_tumor_cells_fpkm)
all_tumor_cells_fpkm <- subset(all_tumor_cells_fpkm,
  select = c(X:RHC6041__Macrophage__.FFFF55)
)


all_nm_cells_fpkm <- read.csv("Data/Single-cell-data/FPKM/GSE81861_CRC_NM_all_cells_FPKM.csv")
rownames(all_nm_cells_fpkm) <- all_nm_cells_fpkm$X
all_nm_cells_fpkm <- gene_name_cleaner(data.to.clean = all_nm_cells_fpkm)
all_nm_cells_fpkm <- t(all_nm_cells_fpkm)
my_tumor_row_names <- rownames(all_nm_cells_fpkm)
all_nm_cells_fpkm <- subset(all_nm_cells_fpkm,
  select = c(X:RHC6187__Macrophage__.FFFF55)
)
```


```{r saving other methods fpkm data prep}
saveRDS(all_tumor_cells_fpkm,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/all_tumor_cells_fpkm.rds"
)
saveRDS(all_nm_cells_fpkm,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/all_nm_cells_fpkm.rds"
)
```


```{r other methods data fpkm prep2}
normal_num <- rep("normal", 266)
tumor_num <- rep("tumor", 375)
all_nums <- c(tumor_num, normal_num)



fpkm_df <- dplyr::bind_cols(all_tumor_cells_fpkm, all_nm_cells_fpkm)
rownames(fpkm_df) <- fpkm_df$X...1
fpkm_df <- subset(fpkm_df, select = c(RHC3546__Tcell__.C6E879:RHC6187__Macrophage__.FFFF55))
fpkm_df$X...377 <- NULL


# Now filtering and pre-processing the data frame
keep_rows <- rowSums(fpkm_df > 5) > 10
table(keep_rows)
fpkm_df <- fpkm_df[keep_rows, ]

my_finished_tumor_rownames <- ifelse(keep_rows == TRUE, my_tumor_row_names,
  print("not in the subsetted list")
)
my_finished_tumor_rownames <- my_finished_tumor_rownames[my_finished_tumor_rownames != "not in the subsetted list"]

fpkm_df <- apply(fpkm_df, c(1, 2), as.numeric)
rownames(fpkm_df) <- my_finished_tumor_rownames
```


```{r other methods data fpkm prep2}
# Looking at the distribution of library sizes
library_size_plot_1 <- ggplot() +
  geom_histogram(aes(x = colSums(fpkm_df)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size for FPKM DF") +
  scale_y_continuous(limits = c(0, 60), expand = c(0, 0))

library_size_plot_1


ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_1_fpkm_df.svg",
  device = "svg", units = "mm", dpi = 600, width = 24, height = 24,
  plot = library_size_plot_1
)



# Normalizing the library size
fpkm_df <- library.size.normalize(t(fpkm_df), verbose = TRUE)
fpkm_df <- t(fpkm_df)
tumor_col_names <- colnames(fpkm_df)
head(as.data.frame(fpkm_df))
fpkm_df <- sqrt(fpkm_df)
head(as.data.frame(fpkm_df))


library_size_plot_2 <- ggplot() +
  geom_histogram(aes(x = colSums(fpkm_df)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size Normalized for FPKM DF") +
  scale_y_continuous(limits = c(0, 35), expand = c(0, 0))

library_size_plot_2

ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_2_fpkm_df.svg",
  device = "svg", units = "mm", dpi = 600, width = 20, height = 20,
  plot = library_size_plot_2
)



# Denoising the single-cell data and saving the output
fpkm_df <- magic_denoiser(
  sc.data = fpkm_df,
  magic.seed = 123,
  magic.solver = "approximate"
)

saveRDS(fpkm_df$denoised_sc_dataframe,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data_fpkm_df.rds"
)
```

## Step 33b: Figure 3b data preprocessing steps for the COUNT varient of the file
```{r other methods count data prep}
# Loading needed functions for tidying the gene names
gene_name_cleaner <- function(data.to.clean = all_tumor_cells_fpkm_denoised_df) {
  data.to.clean <- t(data.to.clean)
  current_colname_split <- strsplit(colnames(data.to.clean), "_")
  finished_gene_list <- c()
  current_list <- current_colname_split
  for (y in seq(1:length(current_list))) {
    finished_gene_list <- c(finished_gene_list, current_list[[y]][2])
  }
  colnames(data.to.clean) <- finished_gene_list
  return(data.to.clean)
}

# Method to clean just a vector of gene names
gene_vector_cleaner <- function(unclean.data = deseq2_rows) {
  split_data <- strsplit(unclean.data, "_")
  finished_genes <- c()
  current_list <- split_data
  for (y in seq(1:length(current_list))) {
    finished_genes <- c(finished_genes, current_list[[y]][2])
  }
  unclean.data <- finished_genes
  return(unclean.data)
}


all_tumor_cells_counts <- read.csv("Data/Single-cell-data/Counts/GSE81861_CRC_tumor_all_cells_COUNT.csv")
rownames(all_tumor_cells_counts) <- all_tumor_cells_counts$X
all_tumor_cells_counts <- gene_name_cleaner(data.to.clean = all_tumor_cells_counts)
all_tumor_cells_counts <- t(all_tumor_cells_counts)
all_tumor_cells_counts <- subset(all_tumor_cells_counts,
  select = c(X:RHC6041__Macrophage__.FFFF55)
)


all_nm_cells_counts <- read.csv("Data/Single-cell-data/Counts/GSE81861_CRC_NM_all_cells_COUNT.csv")
rownames(all_nm_cells_counts) <- all_nm_cells_counts$X
all_nm_cells_counts <- gene_name_cleaner(data.to.clean = all_nm_cells_counts)
all_nm_cells_counts <- t(all_nm_cells_counts)
my_tumor_row_names <- rownames(all_nm_cells_counts)
all_nm_cells_counts <- subset(all_nm_cells_counts,
  select = c(X:RHC6187__Macrophage__.FFFF55)
)


saveRDS(all_tumor_cells_counts,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/all_tumor_cells_counts.rds"
)
saveRDS(all_nm_cells_counts,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/all_nm_cells_counts.rds"
)


normal_num <- rep("normal", 266)
tumor_num <- rep("tumor", 375)
all_nums <- c(tumor_num, normal_num)



counts_df <- dplyr::bind_cols(all_tumor_cells_counts, all_nm_cells_counts)
rownames(counts_df) <- counts_df$X...1
counts_df <- subset(counts_df, select = c(RHC3546__Tcell__.C6E879:RHC6187__Macrophage__.FFFF55))
counts_df$X...377 <- NULL


# Now filtering and pre-processing the data frame
keep_rows <- rowSums(counts_df > 5) > 10
table(keep_rows)
counts_df <- counts_df[keep_rows, ]

my_finished_tumor_rownames <- ifelse(keep_rows == TRUE, my_tumor_row_names,
  print("not in the subsetted list")
)
my_finished_tumor_rownames <- my_finished_tumor_rownames[my_finished_tumor_rownames != "not in the subsetted list"]

counts_df <- apply(counts_df, c(1, 2), as.numeric)
rownames(counts_df) <- my_finished_tumor_rownames


# Looking at the distribution of library sizes
library_size_plot_1 <- ggplot() +
  geom_histogram(aes(x = colSums(counts_df)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size for COUNTS DF") +
  scale_y_continuous(limits = c(0, 60), expand = c(0, 0))

library_size_plot_1


ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_1_counts_df.svg",
  device = "svg", units = "mm", dpi = 600, width = 24, height = 24,
  plot = library_size_plot_1
)



# Normalizing the library size
counts_df <- library.size.normalize(t(counts_df), verbose = TRUE)
counts_df <- t(counts_df)
tumor_col_names <- colnames(counts_df)
head(as.data.frame(counts_df))
counts_df <- sqrt(counts_df)
head(as.data.frame(counts_df))


library_size_plot_2 <- ggplot() +
  geom_histogram(aes(x = colSums(counts_df)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size Normalized for COUNTS DF") +
  scale_y_continuous(limits = c(0, 35), expand = c(0, 0))

library_size_plot_2

ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/single_cell_library_histogram_pre_process_step_2_counts_df.svg",
  device = "svg", units = "mm", dpi = 600, width = 20, height = 20,
  plot = library_size_plot_2
)



# Denoising the single-cell data and saving the output
counts_df <- magic_denoiser(
  sc.data = counts_df,
  magic.seed = 123,
  magic.solver = "approximate"
)

saveRDS(counts_df$denoised_sc_dataframe,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data_counts_df.rds"
)
```

## Step 34: scDD Performance for COAD
```{r scDD performance on COAD}
# For finding the optimal number of genes. Note the scdd file comes from single cell data that is the same for both read and coad
gene_sizes <- seq(100, 3000, 50)
scdd_res <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/scdd_res_read.rds")
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = scdd_res$gene[1:gs],
    my.dataset = "COAD",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/scdd_coefs_coad.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving scDD performance on COAD}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_scdd_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_scdd_top_cindex.rds")
```

## Step 35: DEsingle for COAD
```{r DEsingle on COAD}
# For finding the optimal number of genes
des_results <- readRDS(file = "Data/Reproducible-results/Rmarkdown/Outputs/des_results_read.rds")
des_results <- filter(des_results, pvalue.adj.FDR < 0.05)

gene_sizes <- seq(100, 3000, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(des_results)[1:gs],
    my.dataset = "COAD",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/desingle_coefs_coad.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving DEsingle on COAD}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_desingle_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_desingle_top_cindex.rds")
```

## Step 36: DESeq2 for COAD
```{r deseq2 on COAD}
deseq2_genes <- readRDS(file = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_top_genes.rds")

# For finding the optimal number of genes----
gene_sizes <- seq(100, 1930, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = deseq2_genes[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_coefs_coad.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving deseq2 on COAD}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_deseq2_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_deseq2_top_cindex.rds")
```

## Step 37: edgeR for COAD
```{r edgeR on COAD}
edger_genes <- readRDS(file = "Data/Reproducible-results/Rmarkdown/Outputs/edgeR_1927_genes.rds")

# For finding the optimal number of genes
gene_sizes <- seq(100, 1925, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(edger_genes)[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/edger_coefs_coad.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving edgeR on COAD}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_edger_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_edger_top_cindex.rds")
```


## Step 38: Fig 3a
```{r fig3a data}
# ggplot2 bar chart for the comparison of different methods to our method
edger_coad_max <- readRDS(file = "Data/Reproducible-results/Rmarkdown/Outputs/coad_edger_top_cindex.rds")
deseq2_coad_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/coad_deseq2_top_cindex.rds")
scdd_coad_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/coad_scdd_top_cindex.rds")
des_coad_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/coad_desingle_top_cindex.rds")


coad_other_methods_df <- data.frame(
  method = c(
    "CCsc", "DESeq2",
    "edgeR", "scDD", "DEsingle"
  ),
  c_index = c(
    0.7491, max(deseq2_coad_max),
    max(edger_coad_max),
    max(scdd_coad_max),
    max(deseq2_coad_max)
  )
)
```


```{r saving fig3a data}
saveRDS(coad_other_methods_df, file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_vs_other_methods_df_coad.rds")
write.csv(coad_other_methods_df, file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_vs_other_methods_df_coad.csv")
```

```{r fig3a plot}
color_map <- c("Single-cell" = "#FDE725FF", "Bulk" = "#39568CFF")

coad_other_methods_df$rna_type <- c(
  "Single-cell",
  "Bulk", "Bulk", "Single-cell", "Single-cell"
)



coad_other_methods_df$method <- factor(coad_other_methods_df$method,
  levels = c(
    "CCsc",
    "scDD", "DEsingle",
    "DESeq2",
    "edgeR"
  )
)
cc_singlecell_mms_vs_others_coad <- ggplot(
  data = coad_other_methods_df,
  mapping = aes(
    x = method, y = c_index
  )
) +
  geom_col() +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(face = "bold", hjust = 0.5, size = 32),
    axis.title = element_text(face = "bold", size = 30),
    axis.text = element_text(size = 22),
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    legend.position = "none",
    legend.title = element_text(face = "bold", size = 24),
    legend.text = element_text(size = 22)
  ) +
  xlab("Method") +
  ylab("Mean C-index") +
  ggtitle("TCGA-COAD") +
  coord_cartesian(ylim = c(0.5, 0.85))


cc_singlecell_mms_vs_others_coad
```

```{r saving fig3a plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig3a.svg",
  plot = print(cc_singlecell_mms_vs_others_coad,
    newpage = FALSE
  ),
  device = "svg", dpi = 600,
  width = 5.9, height = 5.9,
  units = "in"
)
```



## Step 39: scDD for READ
```{r scDD performance on TCGA-READ}
condition <- c(
  rep(1, ncol(all_tumor_cells_counts)),
  rep(2, ncol(all_nm_cells_counts))
)

# Removing the two columns represented by X in our original data frames
condition <- condition[2:642]

sce <- SingleCellExperiment(
  assays = list(counts = counts_df$denoised_sc_dataframe),
  colData = data.frame(condition)
)

# Filtering the sce object
sce_filtered <- preprocess(sce, zero.thresh = 0.75, scran_norm = TRUE)

# For multiple cores
BiocParallel::register(BiocParallel::MulticoreParam())

# scDD function and results
prior_param <- list(alpha = 0.01, mu0 = 0, s0 = 0.01, a0 = 0.01, b0 = 0.01)
sce_significance_test <- scDD(sce_filtered,
  prior_param = prior_param,
  testZeroes = FALSE, categorize = FALSE
)

scdd_res <- results(sce_significance_test)
scdd_res <- scdd_res[with(scdd_res, order(nonzero.pvalue.adj)), ]
scdd_res <- filter(scdd_res, nonzero.pvalue.adj < 0.05)
scdd_res <- scdd_res[, c("gene", "nonzero.pvalue", "nonzero.pvalue.adj")]
```


```{r saving scDD performance on TCGA-READ}
saveRDS(scdd_res, file = "Data/Reproducible-results/Rmarkdown/Outputs/scdd_res_read.rds")
write.csv(scdd_res, file = "Data/Reproducible-results/Rmarkdown/Outputs/scdd_res_read.csv")
```

```{r ideal scDD performance on TCGA-READ}
# For finding the optimal number of genes
gene_sizes <- seq(100, 3000, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = scdd_res$gene[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/scdd_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}
```


```{r saving ideal scDD performance on TCGA-READ}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/read_scdd_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/read_scdd_top_cindex.rds")
```


## Step 40: DESeq2 for READ
```{r deseq2 performance on READ}
# Making a summarizedExperiment object from the combined data frame
counts_df_ds2 <- apply(counts_df$denoised_sc_dataframe, c(1, 2), as.integer)

deseq2_se <- SummarizedExperiment(
  assays = list(counts = as.matrix(counts_df_ds2)),
  colData = DataFrame(label = colnames(counts_df$denoised_sc_dataframe)),
  rowData = DataFrame(length = rownames(counts_df$denoised_sc_dataframe))
)
```


```{r}
keep <- rowSums(assay(deseq2_se) > 5) > 10
table(keep)
zinb_data <- deseq2_se[keep, ]

zinb_data$cell_state <- all_nums


zinb_data <- zinb_data[names(zinb_data)[1:dim(zinb_data)[1]], ]

# For multicore
BiocParallel::register(BiocParallel::MulticoreParam())

zinb_data <- zinbwave(zinb_data,
  K = 0, BPPARAM = BiocParallel::bpparam(),
  epsilon = 1e12, normalizedValues = FALSE,
  observationalWeights = TRUE, verbose = TRUE
)

dds <- DESeqDataSet(zinb_data, design = ~cell_state)
dds_red <- DESeqDataSet(zinb_data, design = ~1)

dds <- estimateSizeFactors(dds, type = "poscounts")
dds_red <- estimateSizeFactors(dds_red, type = "poscounts")

scr <- scran::calculateSumFactors(dds)
sizeFactors(dds) <- scr

scr_red <- scran::calculateSumFactors(dds_red)
sizeFactors(dds_red) <- scr_red

dds <- DESeq(dds,
  sfType = "poscounts", useT = TRUE, minmu = 1e-6, minRep = Inf,
  test = "LRT", reduced = ~1
)


dds_res <- DESeq2::results(dds)
dds_res <- dds_res[order(dds_res$padj), ]
deseq2_genes <- rownames(dds_res)
```


```{r saving deseq2 results}
saveRDS(deseq2_genes, file = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_top_genes.rds")
write.csv(deseq2_genes, file = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_top_genes.csv")
```


Now finding the optimal c-index for DESeq2 for READ
```{r optimal gene performance for deseq2 on READ}
# For finding the optimal number of genes
gene_sizes <- seq(100, 1930, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = deseq2_genes[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving optimal gene performance for deseq2 on READ}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/read_deseq2_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/read_deseq2_top_cindex.rds")
```

## Step 41: edgeR for READ
```{r edgeR for READ}
# edgeR
dge <- DGEList(assay(zinb_data))
dge <- calcNormFactors(dge)

weights <- assay(zinb_data, "weights")


design <- model.matrix(~cell_state, data = colData(zinb_data))
dge$weights <- weights
dge <- estimateDisp(dge, design)
fit <- glmFit(dge, design)

lrt <- glmWeightedF(fit, coef = 1:2)
topTags(lrt)

finished_edger <- head(lrt$table, n = 1927)
finished_edger <- finished_edger[order(finished_edger$padjFilter), ]
```


```{r saving edgeR for READ}
write.csv(finished_edger, file = "Data/Reproducible-results/Rmarkdown/Outputs/edgeR_1927_genes.csv")
saveRDS(finished_edger, file = "Data/Reproducible-results/Rmarkdown/Outputs/edgeR_1927_genes.rds")
```

```{r finding ideal gene number for edgeR for READ}
# For finding the optimal number of genes
gene_sizes <- seq(100, 1930, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(finished_edger)[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/edger_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving ideal gene number for edgeR for READ}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/read_edger_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/read_edger_top_cindex.rds")
```

## Step 42: DEsingle for READ
```{r DEsingle for READ}
# Run on lab server using this same code. Modified slightly for server. Code is in file: desingle_server.R
all_tumor_cells_fpkm <- subset(all_tumor_cells_fpkm, select = c(RHC3546__Tcell__.C6E879:RHC6041__Macrophage__.FFFF55))
all_nm_cells_fpkm <- subset(all_nm_cells_fpkm, select = c(RHC3934__Bcell__.7DEA7B:RHC6187__Macrophage__.FFFF55))

condition <- c(rep(1, ncol(all_tumor_cells_fpkm)), rep(2, ncol(all_nm_cells_fpkm)))

condition <- factor(condition)
names(condition) <- c(colnames(all_tumor_cells_fpkm), colnames(all_nm_cells_fpkm))

sce <- SingleCellExperiment(assays = list(counts = deseq2_df$denoised_sc_dataframe))

des_results <- DEsingle(
  counts = sce, group = condition, parallel = TRUE,
  BPPARAM = BiocParallel::bpparam()
)
```


```{r saving DEsingle for READ}
saveRDS(des_results, file = "Data/Reproducible-results/Rmarkdown/Outputs/des_results_read.rds")
write.csv(des_results, file = "Data/Reproducible-results/Rmarkdown/Outputs/des_results_read.csv")

des_results <- filter(des_results, pvalue.adj.FDR < 0.05)
```

Now determing top concordance index for DEsingle on READ
```{r c-index for DEsingle on TCGA-READ}
# For finding the optimal number of genes----
gene_sizes <- seq(100, 5750, 50)
cox_df <- readRDS("Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = cox_df,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(des_results)[1:gs],
    my.dataset = "READ",
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/desingle_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


```{r saving c-index for DEsingle on TCGA-READ}
write.csv(max(my_cindices), file = "Data/Reproducible-results/Rmarkdown/Outputs/read_desingle_top_cindex.csv")
saveRDS(my_cindices, file = "Data/Reproducible-results/Rmarkdown/Outputs/read_desingle_top_cindex.rds")
```


```{r deseq2 filtering}
all_tumor_cells_fpkm <- read.csv("Single-cell-data/Counts/GSE81861_CRC_tumor_all_cells_COUNT.csv")
rownames(all_tumor_cells_fpkm) <- all_tumor_cells_fpkm$X
all_tumor_cells_fpkm <- gene_name_cleaner(data.to.clean = all_tumor_cells_fpkm)
all_tumor_cells_fpkm <- t(all_tumor_cells_fpkm)
all_tumor_cells_fpkm <- subset(all_tumor_cells_fpkm, select = c(X:RHC6041__Macrophage__.FFFF55))


all_nm_cells_fpkm <- read.csv("Single-cell-data/Counts/GSE81861_CRC_NM_all_cells_COUNT.csv")
rownames(all_nm_cells_fpkm) <- all_nm_cells_fpkm$X
all_nm_cells_fpkm <- gene_name_cleaner(data.to.clean = all_nm_cells_fpkm)
all_nm_cells_fpkm <- t(all_nm_cells_fpkm)
my_tumor_row_names <- rownames(all_nm_cells_fpkm)
all_nm_cells_fpkm <- subset(all_nm_cells_fpkm,
  select = c(X:RHC6187__Macrophage__.FFFF55)
)

normal_num <- rep("normal", 266)
tumor_num <- rep("tumor", 375)
all_nums <- c(tumor_num, normal_num)



deseq2_df <- dplyr::bind_cols(all_tumor_cells_fpkm, all_nm_cells_fpkm)
rownames(deseq2_df) <- deseq2_df$X...1
deseq2_df <- subset(deseq2_df, select = c(RHC3546__Tcell__.C6E879:RHC6187__Macrophage__.FFFF55))
deseq2_df$X...377 <- NULL


# Now filtering and pre-processing the data frame
keep_rows <- rowSums(deseq2_df > 5) > 10
table(keep_rows)
deseq2_df <- deseq2_df[keep_rows, ]

my_finished_tumor_rownames <- ifelse(keep_rows == TRUE, my_tumor_row_names, print("not in the subsetted list"))
my_finished_tumor_rownames <- my_finished_tumor_rownames[my_finished_tumor_rownames != "not in the subsetted list"]

deseq2_df <- apply(deseq2_df, c(1, 2), as.numeric)
rownames(deseq2_df) <- tumor_row_names
```


```{r plot of deseq2 filtering}
# Looking at the distribution of library sizes
library_size_plot_1 <- ggplot() +
  geom_histogram(aes(x = colSums(deseq2_df)), bins = 100) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(face = "bold", hjust = 0.5)
  ) +
  xlab("Library Size Distribution") +
  ylab("Count") +
  ggtitle("Single-cell Library Size for DESeq2") +
  scale_y_continuous(limits = c(0, 60), expand = c(0, 0))

library_size_plot_1


ggsave(
  filename = "Outputs/single_cell_library_histogram_pre_process_step_1_deseq2_df.svg",
  device = "svg", units = "mm", dpi = 600, width = 24, height = 24,
  plot = library_size_plot_1
)
```


## Step 43: Fig 3b
```{r fig3b data}
# ggplot2 bar chart for the comparison of different methods to our method
edger_read_max <- readRDS(file = "Data/Reproducible-results/Rmarkdown/Outputs/read_edger_top_cindex.rds")
deseq2_read_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/read_deseq2_top_cindex.rds")
scdd_read_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/read_scdd_top_cindex.rds")
des_read_max <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/des_results_read.rds")


read_other_methods_df <- data.frame(
  method = c(
    "CCsc", "DESeq2",
    "edgeR", "scDD", "DEsingle"
  ),
  c_index = c(
    0.8442, max(deseq2_read_max),
    max(edger_read_max),
    max(scdd_read_max),
    max(deseq2_read_max)
  )
)
```


```{r saving fig3b data}
saveRDS(read_other_methods_df, file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_vs_other_methods_df_read.rds")
write.csv(read_other_methods_df, file = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_vs_other_methods_df_read.csv")
```


```{r fig3b plot}
color_map <- c("Single-cell" = "#FDE725FF", "Bulk" = "#39568CFF")

read_other_methods_df$rna_type <- c(
  "Single-cell", "Bulk",
  "Bulk", "Single-cell", "Single-cell"
)

read_other_methods_df$method <- factor(read_other_methods_df$method,
  levels = c(
    "CCsc", "scDD",
    "DEsingle", "DESeq2",
    "edgeR"
  )
)

cc_singlecell_mms_vs_others_read <- ggplot(
  data = read_other_methods_df,
  mapping = aes(
    x = method, y = c_index
  )
) +
  geom_col() +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(face = "bold", hjust = 0.5, size = 32),
    axis.title = element_text(face = "bold", size = 30),
    axis.text.x = element_text(size = 22, angle = 90),
    axis.text.y = element_text(size = 22),
    legend.position = "none",
    legend.title = element_text(face = "bold", size = 24),
    legend.text = element_text(size = 22)
  ) +
  xlab("Method") +
  ylab(NULL) +
  ggtitle("TCGA-READ") +
  coord_cartesian(ylim = c(0.5, 0.85))

cc_singlecell_mms_vs_others_read
```


```{r saving fig3b}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig3b.svg",
  plot = print(cc_singlecell_mms_vs_others_read,
    newpage = FALSE
  ),
  device = "svg", dpi = 600,
  width = 5.9, height = 5.9,
  units = "in"
)
```
## Combined Fig3
```{r combined fig3}
fig3 <- ggarrange(cc_singlecell_mms_vs_others_coad, cc_singlecell_mms_vs_others_read,
  nrow = 1,
  ncol = 2,
  common.legend = TRUE, legend = "none"
)
```

```{r saving fig3 combo plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig3.svg",
  plot = print(fig3, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.48, height = 7.48,
  units = "in"
)
```


## Step 44a: Fig 4b data
```{r fig4b}
read_coef_df <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_read_coefs_at_top_combo_and_index.csv")
read_coef_df <- read_coef_df[, 2:3]
colnames(read_coef_df) <- c("Gene", "coefs")
read_coef_df$hazard_ratio <- exp(read_coef_df$coefs)

# Doing analysis for those genes that have hazard ratios < 0.5 and
# a separate analysis for genes that have hazard ratios > 2
read_coef_df_sub_pos <- filter(read_coef_df, hazard_ratio < 0.5)
read_coef_df_sub_neg <- filter(read_coef_df, hazard_ratio > 2)
read_coef_df_sub_pos$log_hazard_ratio <- log2(read_coef_df_sub_pos$hazard_ratio)
read_coef_df_sub_neg$log_hazard_ratio <- log2(read_coef_df_sub_neg$hazard_ratio)

# TRIM60 is removed because it has an infinite hazard ratio and
# therefore cannot be plotted
read_coef_df_sub_pos <- filter(read_coef_df_sub_pos, Gene != "TRIM60")
read_coef_df_sub_neg <- filter(read_coef_df_sub_neg, Gene != "TRIM60")
```

## Step 44b: Fig 4b increasing risk plot
```{r fig4b (cc singlecell mms READ coef plot)}
level_order <- order(read_coef_df_sub_neg$log_hazard_ratio, decreasing = TRUE)
read_coef_df_sub_neg_ordered <- read_coef_df_sub_neg[level_order, ]
read_coef_plot <- ggplot(data = read_coef_df_sub_neg, aes(
  x = factor(Gene, levels = rev(c("GRAMD4P2", "ENPP7P6", "UMODL1", "ZNF425", "PPIAP16", "FMO1", "LGSN", "DGKB", "RN7SKP203", "MFAP3L", "PLXNA4", "DUSP26", "C4A", "MAP3K6", "CD274", "PCMTD1P3"))), y = log_hazard_ratio, color = "red", fill = "red"
)) +
  geom_col(color = "red", fill = "red") +
  theme_bw() +
  ggtitle("Top Increasing") +
  ylab(bquote("log"[2] ~ "(Hazard Ratio)")) +
  xlab(NULL) +
  theme(
    text = element_text(family = "Arial"),
    plot.title = element_text(
      hjust = 0.5, face = "bold", size = 32
    ),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  coord_flip() +
  scale_y_break(breaks = c(35, 250), expand = expansion(mult = c(0, 0.1)))


read_coef_plot
```

## Step 44c: Saving fig 4b plot
```{r saving fig 4b (cc singlecell mms READ coef plot)}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig4b.svg",
  plot = print(read_coef_plot),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


```{r fig4d pos (cc singlecell mms READ coef plot)}
level_order <- order(read_coef_df_sub_pos$log_hazard_ratio, decreasing = FALSE)
read_coef_df_sub_pos_ordered <- read_coef_df_sub_pos[level_order, ]
read_coef_plot_pos <- ggplot(data = read_coef_df_sub_pos, aes(
  x = factor(Gene, levels = rev(c("WDR87", "KCNJ6", "FAM182A", "GCNT1P1", "HNRNPA1P40", "ZNF23", "SLC2A14", "PCDH11X", "RASSF9", "MBLAC2", "RGS12", "ACAN", "SLC35F1", "PKHD1", "C17orf77", "MEG3", "KCNQ3", "DAP3P2", "SLC44A5"))), y = log_hazard_ratio,
  color = "pale green", fill = "pale green"
)) +
  geom_col(color = "pale green", fill = "pale green") +
  theme_bw() +
  ggtitle("Top Decreasing") +
  ylab(bquote("log"[2] ~ "(Hazard Ratio)")) +
  xlab("Genes") +
  theme(
    legend.position = "none",
    text = element_text(family = "Arial"),
    plot.title = element_text(
      hjust = 0.5, face = "bold", size = 32
    ),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  coord_flip() +
  scale_y_break(breaks = c(-50, -80), expand = expansion(mult = c(0.05, 0)), scales = 0.5)


read_coef_plot_pos
```

```{r saving fig 4b pos (cc singlecell mms READ coef plot)}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/fig4b_pos.svg",
  plot = print(read_coef_plot_pos),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


# Supplemental figures

## Supplement S1a: miRNA heatmap metric 
```{r figs1a}
mirna_genes <- mirna_calculator(print.ts.targets = TRUE, max.mirnas = 10, max.miR.targets = 10)

figs1a <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/figs1a_smaller_matrix.rds")

# figs1a <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/figs1_mirna_score_matrix.rds")


svg(filename = "~/Desktop/figs1a.svg", width = 8, height = 8)
figs1a_p <- heatmap(as.matrix(figs1a), Rowv = NA, Colv = NA, main = "miRNA Interaction Map", margins = c(10, 5), xlab = "miRNAs", ylab = "Genes", col = c("#ffffbf", "#ef8a62"))
dev.off()
```

```{r figs1b}
figs1b <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/figs1_mirnas.rds")
figs1b <- as.data.frame(figs1b)
colnames(figs1b) <- "score"
figs1b$gene <- rownames(figs1b)
figs1b_sub <- head(figs1b)
figs1b_plot <- ggplot(data = figs1b_sub, aes(x = reorder(gene, -score, sum), y = score, color = gene, fill = gene)) +
  geom_col() +
  theme(
    panel.background = element_blank(),
    legend.position = "none",
    plot.title = element_text(face = "bold", size = 32, hjust = 0.5),
    axis.title = element_text(face = "bold", size = 30),
    axis.text.x = element_text(size = 22, angle = 90, vjust = 0.5),
    axis.text.y = element_text(size = 22)
  ) +
  xlab("Gene") +
  ylab("Score") +
  ggtitle("miRNA Metric Score")
```

```{r saving figs1b}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs1b.svg",
  plot = print(figs1b_plot, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7, height = 7,
  units = "in"
)
```


## Step 45a: Supplement S2a: Heatmaps of COAD miRNA-miRNA target spaces
```{r fig S2a data}
# COAD Heatmap
combo_used <- c(
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010",
  "600_10", "600_110", "600_210", "600_310", "600_410", "600_510", "600_610", "600_710", "600_810", "600_910", "600_1010",
  "500_10", "500_110", "500_210", "500_310", "500_410", "500_510", "500_610", "500_710", "500_810", "500_910", "500_1010",
  "400_10", "400_110", "400_210", "400_310", "400_410", "400_510", "400_610", "400_710", "400_810", "400_910", "400_1010",
  "300_10", "300_110", "300_210", "300_310", "300_410", "300_510", "300_610", "300_710", "300_810", "300_910", "300_1010",
  "200_10", "200_110", "200_210", "200_310", "200_410", "200_510", "200_610", "200_710", "200_810", "200_910", "200_1010",
  "100_10", "100_110", "100_210", "100_310", "100_410", "100_510", "100_610", "100_710", "100_810", "100_910", "100_1010"
)

figs2a_df <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_coad_top_cindex.csv")
colnames(figs2a_df) <- c("combo_used", "c_index")
figs2a_df$combo_used <- combo_used
figs2a_df$mirna_num <- rep(seq(800, 100, -100), each = 11)
figs2a_df$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)
```

## Step 45b: S2a plot
```{r fig S2a plot}
heatmap_figs2a <- ggplot(data = figs2a_df, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white", size = 3, fontface = "bold", family = "Arial") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS COAD",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    axis.title.x = element_text(size = 20, face = "bold"),
    axis.title.y = element_text(size = 20, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 20),
    legend.spacing.x = unit(0.8, "cm"),
    legend.position = "bottom",
    legend.key.width = unit(1.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_figs2a_finished <- heatmap_figs2a + scale_fill_viridis_c()
heatmap_figs2a_finished
```

## Step 45c: Saving S2a plot
```{r saving fig S2a plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs2a.svg",
  device = "svg", units = "in", dpi = 600, width = 7.8, height = 7.8,
  plot = heatmap_figs2a_finished
)
```

## Step 45d: S2b data
```{r fig S2b data}
# READ Heatmap. Had to go back and get the 700 and 600 miRNA data range because
# I forgot to include it in the initial pass. Hence why I load both of my raw
# data frames here. They were combined in excel into a finished data frame.
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010",
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010",
  "600_10", "600_110", "600_210", "600_310", "600_410", "600_510", "600_610", "600_710", "600_810", "600_910", "600_1010",
  "500_10", "500_110", "500_210", "500_310", "500_410", "500_510", "500_610", "500_710", "500_810", "500_910", "500_1010",
  "400_10", "400_110", "400_210", "400_310", "400_410", "400_510", "400_610", "400_710", "400_810", "400_910", "400_1010",
  "300_10", "300_110", "300_210", "300_310", "300_410", "300_510", "300_610", "300_710", "300_810", "300_910", "300_1010",
  "200_10", "200_110", "200_210", "200_310", "200_410", "200_510", "200_610", "200_710", "200_810", "200_910", "200_1010",
  "100_10", "100_110", "100_210", "100_310", "100_410", "100_510", "100_610", "100_710", "100_810", "100_910", "100_1010"
)

# figs2b_df_initial_df <- read.csv("~/Desktop/All-files-from-paper-draft-update/cc_singlecell_mms_read_top_cindex.csv")
# figs2b_df_continued_df <- read.csv("~/Desktop/All-files-from-paper-draft-update/cc_singlecell_mms_read_top_cindex_800_700_mirna_range.csv")
# colnames(figs3b_df) <- c("combo_used", "c_index")
# figs2b_df$combo_used <- combo_used
# write.csv(figs3b_df, file("~/Desktop/All-files-from-paper-draft-update/read_grid_search_data.csv"))
figs2b_df_done <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/read_grid_search_data.csv")
figs2b_df_done$mirna_num <- rep(seq(1000, 100, -100), each = 11)
figs2b_df_done$mirna_target <- rep(seq(10, 1010, by = 100), times = 10)
```

## Step 45e: S2b plot
```{r fig S2b plot}
heatmap_figs2b <- ggplot(data = figs2b_df_done, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white", size = 3, fontface = "bold", family = "Arial") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS READ",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    axis.title.x = element_text(size = 20, face = "bold"),
    axis.title.y = element_text(size = 20, face = "bold"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 20),
    legend.position = "bottom",
    legend.spacing.x = unit(0.8, "cm"),
    legend.key.width = unit(1.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_figs2b_finished <- heatmap_figs2b + scale_fill_viridis_c()
heatmap_figs2b_finished
```
## Step 45f: Saving S2b plot
```{r saving fig S2b plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs2b.svg",
  device = "svg", units = "in", dpi = 600, width = 7.8, height = 7.8,
  plot = heatmap_figs2b_finished
)
```

## Step 46a: Supplement S3a: Active gene size graph
```{r fig S3a coad active gene data}
coad_active_coef_df <- data.frame(
  coef_num = c(191, 255, 211, 208, 184),
  method = c(
    "CCsc", "scDD",
    "DEsingle", "DESeq2", "edgeR"
  )
)
```

## Step 46b: S3a Plot
```{r fig S3a coad active gene plot}
color_map <- c("Our Method" = "#FDE725FF", "Other Method" = "#39568CFF")
coad_active_coef_df$method_type <- c(
  "Our Method", "Other Method",
  "Other Method", "Other Method",
  "Other Method"
)

coad_active_gene_plot <- ggplot(
  data = coad_active_coef_df,
  aes(
    x = method, y = coef_num, fill = method_type
  )
) +
  geom_col() +
  theme_bw() +
  ggtitle("COAD") +
  ylab("Active Gene Set Size") +
  xlab("Methods") +
  theme(
    legend.position = "none",
    text = element_text(family = "Arial"),
    plot.title = element_text(
      hjust = 0.5, face = "bold", size = 32
    ),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 22),
    axis.text.y = element_text(size = 22),
    legend.text = element_text(size = 22),
    legend.title = element_text(size = 30)
  ) +
  scale_color_viridis_d(direction = -1) +
  scale_fill_viridis_d(direction = -1) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  coord_flip() +
  scale_fill_manual(values = color_map, name = "Method Type")

coad_active_gene_plot
```

## Step 46c: Saving S3a plot
```{r fig S3a saving coad active gene plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs3a.svg",
  plot = print(coad_active_gene_plot, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```

## Step 46d: S3b data
```{r fig S3b read active gene data}
read_active_coef_df <- data.frame(
  coef_num = c(52, 6, 4, 59, 61),
  method = c(
    "CCsc", "scDD",
    "DEsingle", "DESeq2", "edgeR"
  )
)
```

## Step 46e: S3b plot
```{r fig S3b read active gene plot}
color_map <- c("Our Method" = "#FDE725FF", "Other Method" = "#39568CFF")
read_active_coef_df$method_type <- c(
  "Our Method", "Other Method",
  "Other Method", "Other Method",
  "Other Method"
)

read_active_gene_plot <- ggplot(
  data = read_active_coef_df,
  aes(
    x = method, y = coef_num,
    fill = method_type
  )
) +
  geom_col() +
  theme_bw() +
  ggtitle("READ") +
  ylab("Active Gene Set Size") +
  xlab("Methods") +
  theme(
    legend.position = "none",
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    axis.text.x = element_text(size = 18),
    axis.text.y = element_text(size = 18)
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  coord_flip() +
  scale_fill_manual(values = color_map, name = "Method Type")

read_active_gene_plot
```

## Step 46f: Saving S3b plot
```{r fig S3b saving active read gene plot}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs3b.svg",
  plot = print(read_active_gene_plot, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


##S4a (making and saving)
```{r barchart of cc singlecell performance on non tcga crc}
color_map <- c("Our Method" = "#FDE725FF", "Other Method" = "#39568CFF")
df <- data.frame(method = c("CCsc MM", "CCsc MS", "CCsc MMS", "MAD", "SDE", "miRNA", "MAD + SDE", "Random Genes"), cindex = c(0.7459, 0.7502, 0.7911, 0.5634, 0.5, 0.6745, 0.5746, 0.59857))

df$method <- factor(df$method, levels = c("CCsc MM", "CCsc MS", "CCsc MMS", "MAD", "SDE", "miRNA", "MAD + SDE", "Random Genes"))

df$method_type <- c("Our Method", "Our Method", "Our Method", "Other Method", "Other Method", "Other Method", "Other Method", "Other Method")

non_tcga_crc_p <- ggplot(data = df, aes(x = method, y = cindex, fill = method_type)) +
  geom_col() +
  theme(
    text = element_text(family = "Arial"),
    axis.text.x = element_text(size = 18, angle = 90, vjust = 0.5),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold", hjust = 0.5),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    legend.position = "none",
    panel.background = element_blank()
  ) +
  coord_cartesian(ylim = c(0.5, 0.8)) +
  ggtitle("Non-TCGA CRC") +
  xlab("Method") +
  ylab("Mean C-index") +
  scale_fill_manual(values = color_map, name = "Method Type")

non_tcga_crc_p <- non_tcga_crc_p +
  scale_x_discrete(labels = function(x) str_wrap(x, width = 6))

ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs4a.svg",
  plot = print(non_tcga_crc_p, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


## S4b (making and saving)
```{r plot of cc singlecell mms vs. other methods non tcga crc}
color_map <- c("Our Method" = "#FDE725FF", "Other Method" = "#39568CFF")
df <- data.frame(
  method = c("CCsc", "DEsingle", "scDD", "DESeq2", "edgeR"),
  cindex = c(0.7911, 0.5625, 0.6364, 0.5667, 0.702)
)

df$method <- factor(df$method, levels = c("CCsc", "edgeR", "scDD", "DESeq2", "DEsingle"))

df$method_type <- c("Our Method", "Other Method", "Other Method", "Other Method", "Other Method")

non_tcga_crc_p <- ggplot(data = df, aes(x = method, y = cindex, fill = method_type)) +
  geom_col() +
  theme(
    text = element_text(family = "Arial"),
    axis.text.x = element_text(size = 18, angle = 90, vjust = 0.5),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold", hjust = 0.5),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    legend.position = "none",
    panel.background = element_blank()
  ) +
  coord_cartesian(ylim = c(0.5, 0.8)) +
  ggtitle("Non-TCGA CRC") +
  xlab("Method") +
  ylab("Mean C-index") +
  scale_fill_manual(values = color_map, name = "Method Type")

non_tcga_crc_p <- non_tcga_crc_p +
  scale_x_discrete(labels = function(x) str_wrap(x, width = 10))


ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs4b.svg",
  plot = print(non_tcga_crc_p, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


## Fig S5: miRNA suppression analysis: Genes of signature expression vs. mean expression of all genes
```{r figs5 coad}
# COAD
cox_df <- readRDS(file = "/Users/andrewwillems/Documents/Work/PhD_Program/Hong_Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

cox_df <- cox_df[, 1:55150]

cox_df <- apply(cox_df, 2, as.numeric)


# Mean of each gene
mean_gene_expression_per_gene <- apply(cox_df, 2, mean)


# All are well below the mean
med_gene_expression_all <- median(cox_df) # 0.01830062
mean_gene_expression_all <- mean(cox_df) # 4.91491

mean_gene_expression_per_gene["ZNF705D"] # 3.693398e-05
mean_gene_expression_per_gene["UNC5D"] # 0.01910989
mean_gene_expression_per_gene["TP53TG3D"] # 0.005964462
mean_gene_expression_per_gene["ST6GALNAC3"] # 0.312284
mean_gene_expression_per_gene["RSPH10B"] # 0.01371021
mean_gene_expression_per_gene["KCNC1"] # 0.01533954
mean_gene_expression_per_gene["HS6ST3"] # 0.03071341
mean_gene_expression_per_gene["FAM182A"] # 0.01491377
mean_gene_expression_per_gene["FABP7"] # 0.04177722
mean_gene_expression_per_gene["DNAJC5G"] # 0.02267668
mean_gene_expression_per_gene["ABCA13"] # 0.09333975
mean_gene_expression_per_gene["OPCML"] # 0.06226365
mean_gene_expression_per_gene["RFPL3S"] # 0.08692278

mean_df <- data.frame(genes = c("ZNF705D", "UNC5D", "TP53TG3D", "ST6GALNAC3", "RSPH10B", "KCNC1", "HS6ST3", "FAM182A", "FABP7", "DNAJC5G", "ABCA13", "OPCML", "RFPL3S"), mean_expr = c(0.00003693398, 0.01910989, 0.005964462, 0.312284, 0.01371021, 0.01533954, 0.03071341, 0.01491377, 0.04177722, 0.02267668, 0.09333975, 0.06226365, 0.08692278))

figs5a <- ggplot(data = mean_df, aes(x = genes, y = mean_expr)) +
  theme(
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    panel.background = element_blank(),
    axis.text.x = element_text(size = 18, angle = 90),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold")
  ) +
  xlab("Genes") +
  ylab("Mean Expression Level") +
  ggtitle("Increasing") +
  geom_col() +
  geom_hline(yintercept = mean_gene_expression_all, color = "red")

figs5a
```

```{r saving figs5 coad}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs5a_increase.svg",
  plot = print(figs5a, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```



```{r figs5 read}
# READ
cox_df <- readRDS(file = "/Users/andrewwillems/Documents/Work/PhD_Program/Hong_Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")

cox_df <- cox_df[, 1:55150]

cox_df <- apply(cox_df, 2, as.numeric)


# Mean of each gene
mean_gene_expression_per_gene <- apply(cox_df, 2, mean)


# Mean of all genes in the set. All but one gene from our signatures  expression level is lower than the total gene set median and all are well below the mean
med_gene_expression_all <- median(cox_df) # 0.01692303
mean_gene_expression_all <- mean(cox_df) # 4.801431

mean_gene_expression_per_gene["ZNF425"] # 0.5764706
mean_gene_expression_per_gene["UMODL1"] # 0.2066285
mean_gene_expression_per_gene["RN7SKP203"] # 0.01459431
mean_gene_expression_per_gene["PPIAP16"] # 0.1908489
mean_gene_expression_per_gene["PLXNA4"] # 0.2265823
mean_gene_expression_per_gene["MFAP3L"] # 0.7426766
mean_gene_expression_per_gene["LGSN"] # 0.1761296
mean_gene_expression_per_gene["GRAMD4P2"] # 0.001507944
mean_gene_expression_per_gene["FMO1"] # 0.2076086
mean_gene_expression_per_gene["ENPP7P6"] # 0.01471783
mean_gene_expression_per_gene["DUSP26"] # 0.1764655
mean_gene_expression_per_gene["DGKB"] # 0.1037039
mean_gene_expression_per_gene["MAP3K6"] # 4.367673
mean_gene_expression_per_gene["CD274"] # 0.9265985
mean_gene_expression_per_gene["C4A"] # 0.5138211
mean_gene_expression_per_gene["PCMTD1P3"] # 0.1842997

mean_df <- data.frame(genes = c("ZNF425", "UMODL1", "RN7SKP203", "PPIAP16", "PLXNA4", "MFAP3L", "LGSN", "GRAMD4P2", "FMO1", "ENPP7P6", "DUSP26", "DGKB", "MAP3K6", "CD274", "C4A", "PCMTD1P3"), mean_expr = c(0.5764706, 0.2066285, 0.01459431, 0.1908489, 0.2265823, 0.7426766, 0.1761296, 0.001507944, 0.2076086, 0.01471783, 0.1764655, 0.1037039, 4.367673, 0.9265985, 0.5138211, 0.1842997))

figs5b <- ggplot(data = mean_df, aes(x = genes, y = mean_expr)) +
  theme(
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    panel.background = element_blank(),
    axis.text.x = element_text(size = 18, angle = 90),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold")
  ) +
  xlab("Genes") +
  ylab("Mean Expression Level") +
  ggtitle("Increasing") +
  geom_col() +
  geom_hline(yintercept = mean_gene_expression_all, color = "red")

figs5b
```

```{r saving figs5 read}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs5b_increase.svg",
  plot = print(figs5b, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```

```{r figs5c coad}
# COAD
cox_df <- readRDS(file = "/Users/andrewwillems/Documents/Work/PhD_Program/Hong_Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Data/coad_df_finished.rds")

cox_df <- cox_df[, 1:55150]

cox_df <- apply(cox_df, 2, as.numeric)


# Mean of each gene
mean_gene_expression_per_gene <- apply(cox_df, 2, mean)


# All are well below the mean
med_gene_expression_all <- median(cox_df) # 0.01830062
mean_gene_expression_all <- mean(cox_df) # 4.801431

mean_gene_expression_per_gene["AJAP1"] # 0.04617215
mean_gene_expression_per_gene["SLC24A5"] # 0.004460766
mean_gene_expression_per_gene["CACNG8"] # 0.1284229
mean_gene_expression_per_gene["C1orf229"] # 0.09790357
mean_gene_expression_per_gene["GRAMD4P3"] # 0.001548422
mean_gene_expression_per_gene["ICOS"] # 0.6582565
mean_gene_expression_per_gene["HAP1"] # 0.09382534
mean_gene_expression_per_gene["TENM2"] # 0.06206247
mean_gene_expression_per_gene["AC079612.1"] # 0.01333948
mean_gene_expression_per_gene["AL133373.1"] # 0.1261812
mean_gene_expression_per_gene["BSND"] # 0.008760493
mean_gene_expression_per_gene["RS1"] # 0.02541751

mean_df <- data.frame(genes = c("AJAP1", "SLC24A5", "CACNG8", "C1orf229", "GRAMD4P3", "ICOS", "HAP1", "TENM2", "AC079612.1", "AL133373.1", "BSND", "RS1"), mean_expr = c(0.04617215, 0.004460766, 0.1284229, 0.09790357, 0.001548422, 0.6582565, 0.09382534, 0.06206247, 0.01333948, 0.1261812, 0.008760493, 0.02541751))

figs5c <- ggplot(data = mean_df, aes(x = genes, y = mean_expr)) +
  theme(
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    panel.background = element_blank(),
    axis.text.x = element_text(size = 18, angle = 90),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold")
  ) +
  xlab("Genes") +
  ylab("Mean Expression Level") +
  ggtitle("Decreasing") +
  geom_col() +
  geom_hline(yintercept = mean_gene_expression_all, color = "red")

figs5c
```
```{r}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs5a_decrease.svg",
  plot = print(figs5c, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```




```{r figs5d}
cox_df <- readRDS(file = "/Users/andrewwillems/Documents/Work/PhD_Program/Hong_Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Data/read_df_finished.rds")

cox_df <- cox_df[, 1:55150]

cox_df <- apply(cox_df, 2, as.numeric)


# Mean of each gene
mean_gene_expression_per_gene <- apply(cox_df, 2, mean)


# Mean of all genes in the set. All but one gene from our signatures  expression level is lower than the total gene set median and all are well below the mean
med_gene_expression_all <- median(cox_df) # 0.01692303
mean_gene_expression_all <- mean(cox_df) # 4.801431


mean_gene_expression_per_gene["FAM182A"] # 0.01214589
mean_gene_expression_per_gene["SLC35F1"] # 0.1814575
mean_gene_expression_per_gene["RGS12"] # 1.722252
mean_gene_expression_per_gene["PKHD1"] # 0.0708791
mean_gene_expression_per_gene["GCNT1P1"] # 0.01056198
mean_gene_expression_per_gene["C17orf77"] # 0.9222566
mean_gene_expression_per_gene["KCNJ6"] # 0.007430623
mean_gene_expression_per_gene["RASSF9"] # 0.4069409
mean_gene_expression_per_gene["DAP3P2"] # 0.2036437
mean_gene_expression_per_gene["WDR87"] # 0.003090666
mean_gene_expression_per_gene["KCNQ3"] # 0.1142717
mean_gene_expression_per_gene["PCDH11X"] # 0.002465849
mean_gene_expression_per_gene["MEG3"] # 0.8921096
mean_gene_expression_per_gene["MBLAC2"] # 1.557469
mean_gene_expression_per_gene["SLC44A5"] # 1.034524
mean_gene_expression_per_gene["HNRNPA1P40"] # 0.0356591
mean_gene_expression_per_gene["ZNF23"] # 0.1588508
mean_gene_expression_per_gene["ACAN"] # 1.049572
mean_gene_expression_per_gene["SLC2A14"] # 0.07458324


mean_df <- data.frame(genes = read_coef_df_sub_pos$Gene, mean_expr = c(0.01214589, 0.1814575, 1.722252, 0.0708791, 0.01056198, 0.9222566, 0.007430623, 0.4069409, 0.2036437, 0.003090666, 0.1142717, 0.002465849, 0.8921096, 1.557469, 1.034524, 0.0356591, 0.1588508, 1.049572, 0.07458324))

figs5d <- ggplot(data = mean_df, aes(x = genes, y = mean_expr)) +
  theme(
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 32),
    panel.background = element_blank(),
    axis.text.x = element_text(size = 18, angle = 90),
    axis.text.y = element_text(size = 18),
    axis.title = element_text(size = 30, face = "bold")
  ) +
  xlab("Genes") +
  ylab("Mean Expression Level") +
  ggtitle("Decreasing") +
  geom_col() +
  geom_hline(yintercept = mean_gene_expression_all, color = "red")

figs5d
```
```{r saving figs5 read}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/figs5b_decrease.svg",
  plot = print(figs5d, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```


## Fig S4 Analysis Additional bulk CRC datasets for validation of our method that is non-TCGA
```{r non-tcga bulk crc dataset}
patient_meta <- read.csv("Data/Reproducible-results/Rmarkdown/Data/rectal_msk_2022/data_clinical_patient.txt", sep = "\t")

patient_meta <- subset(patient_meta, select = c("X.Patient.Identifier", "Has.WES", "Overall.Survival.Neo", "Overall.Survival.Status"))

patient_meta <- filter(patient_meta, Overall.Survival.Status != "NA")
patient_meta <- slice(patient_meta, 5:n())
patient_meta <- filter(patient_meta, Has.WES == 1)
patient_meta$Overall.Survival.Status <- ifelse(patient_meta$Overall.Survival.Status == "0:LIVING", 0, 1)
colnames(patient_meta)[1] <- "ID"
colnames(patient_meta)[3:4] <- c("days.to.last.follow.up", "vital.status")
patient_meta$ID <- gsub(pattern = "-", replacement = ".", x = patient_meta$ID)
patient_meta$days.to.last.follow.up <- as.numeric(patient_meta$days.to.last.follow.up)
patient_meta$days.to.last.follow.up <- patient_meta$days.to.last.follow.up * 30
patient_meta <- na.omit(patient_meta)

patient_rna <- read.csv("Data/Reproducible-results/Rmarkdown/Data/rectal_msk_2022/data_mrna_seq_expression.txt", sep = "\t")

hugo_rna <- patient_rna$Hugo_Symbol

patient_rna <- subset(patient_rna, select = grepl("WES", colnames(patient_rna)))
patient_rna$gene <- hugo_rna
colnames(patient_rna) <- gsub(pattern = ".T01.WES", replacement = "", x = colnames(patient_rna))
common_patients <- intersect(colnames(patient_rna), patient_meta$ID)

patient_meta <- filter(patient_meta, ID %in% colnames(patient_rna))
patient_rna <- t(patient_rna)
colnames(patient_rna) <- patient_rna["gene", ]
patient_rna <- patient_rna[-83, ]
patient_rna <- as.data.frame(patient_rna)
patient_rna <- patient_rna[, !duplicated(colnames(patient_rna))]
# patient_rna <- filter(patient_rna, colnames(patient_rna) %in% patient_meta$ID)
patient_rna$ID <- row.names(patient_rna)
patient_df <- merge(patient_rna, patient_meta, by = "ID")
row.names(patient_df) <- patient_df$ID
patient_df <- patient_df[, 2:19532]

non_tcga_cox <- patient_df
```

```{r ideal gene size for mad on non-tcga-crc-cancer}
# Getting ideal gene number for MAD metric on non TCGA cancer
gene_sizes <- seq(100, 3000, 50)
mad_cindices <- rep(0, length(gene_sizes))

# Loading in the MAD file if they aren't already loaded into the
# environment from earlier
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Data/mad_colon_and_rectal_cancer.rds")


for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = mad.genes,
    cox.df = non_tcga_cox,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    n.folds = 10,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/mad_non_tcga_crc_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mad_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mad_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mad_cindices_read_df <- as.data.frame(cbind(gene_sizes, mad_cindices))
mad_cindices_read_df$method <- rep("MAD", nrow(mad_cindices_read_df))
colnames(mad_cindices_read_df)[2] <- "c_index"
write.csv(mad_cindices_read_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/mad_cindices_non_tcga_crc_across_gene_size.csv"
)
```

```{r ideal gene size for sde on non-tcga-crc-cancer}
# Getting ideal gene number for SDE metric on non-tcga cancer
gene_sizes <- seq(100, 3000, 50)
sde_cindices <- rep(0, length(gene_sizes))

sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Data/sde_colon_and_rectal_cancer.rds")

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = sde.genes,
    cox.df = non_tcga_cox,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    n.folds = 10,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/sde_non_tcga_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the sde_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  sde_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
sde_cindices_read_df <- as.data.frame(cbind(gene_sizes, sde_cindices))
sde_cindices_read_df$method <- rep("SDE", nrow(sde_cindices_read_df))
colnames(sde_cindices_read_df)[2] <- "c_index"
write.csv(sde_cindices_read_df,
  file = "Data/Reproducible-results/Rmarkdown/Outputs/sde_cindices_non_tcga_across_gene_size.csv"
)
```

```{r getting ideal MAD + SDE gene performance on non-tcga-crc}
# Top performance is: 0.5958
mad.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde.genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")


combo_used <- "1000_1010"

alpha_value <- c(0.0, 0.5, 1.0)

gene_sizes <- seq(100, 3000, 50)
data_set <- "non_tcga_crc"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[11]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    mad_sde_optimized <- two_weight_optimizer(
      first.metric = mad.genes,
      second.metric = sde.genes,
      my.filename = paste0("Data/Reproducible-results/Rmarkdown/Optimizations/Optimization_", c, "_targets_mad_sde", data_set, ".rds")
    )



    index <- 1

    for (md in mad_sde_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          cox.predictors = md,
          cox.df = non_tcga_cox,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          n.folds = 10,
          my.filename = paste0("mad_sde_non_tcga_coefs.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
        index <- index + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
      index <- index + 1
    }

    write.csv(my_cindices, file = paste0("mad_sde_non_tcga_df.csv"))
  }
}
```


```{r ideal gene number for high miRNA-miRNA target size on non-tcga-crc}
# Getting ideal gene number for miRNA metric on TCGA-READ
# For this metric we don't know which combination of miRNA and miRNA targets
# will yield the best result in advance so we are trying 3 different miRNA-
# miRNA target combinations that encompass specific areas of our grid search.
# They are low miRNA number and low miRNA target number, medium miRNA number and
# medium miRNA target number, and high miRNA number and high miRNA target number.
# Once the ideal miRNA-miRNA target pair is known from the grid search we will
# also include a gene size search for it here. The top performer is 0.7512
gene_sizes <- seq(100, 3000, 50)
mirna_high_cindices <- rep(0, length(gene_sizes))
mirna_medium_cindices <- rep(0, length(gene_sizes))
mirna_low_cindices <- rep(0, length(gene_sizes))

# Loading the high miRNA-miRNA target file
# load("Data/Reproducible-results/Rmarkdown/Data/800_1010_targets.RData",
#      verbose = TRUE)
# load("Data/Reproducible-results/Data/1000_110_targets.RData", verbose = TRUE)
load("Data/Reproducible-results/Data/1000_1010_targets.RData", verbose = TRUE)

high.mirna.genes <- mirna.ranking

for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = high.mirna.genes,
    cox.df = non_tcga_cox,
    gene.num = gs,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    n.folds = 10,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/high_mirna_non_tcga_coefs_", gs, "_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the mirna_high_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  mirna_high_cindices[which(gene_sizes == gs)] <- top_cindex
}

# Binding the gene size and c-index vectors together to get the finished data
# frame
mirna_high_cindices_read_df <- as.data.frame(cbind(
  gene_sizes,
  mirna_high_cindices
))

colnames(mirna_high_cindices_read_df)[2] <- "c_index"
mirna_high_cindices_read_df$mirna_type <- rep(
  "high",
  nrow(mirna_high_cindices_read_df)
)

write.csv(
  mirna_high_cindices_read_df,
  "Data/Reproducible-results/Rmarkdown/Outputs/mirna_high_cindices_non_tcga_across_gene_size.csv"
)
```
## Random gene cindex is: 0.59857 
```{r random genes non tcga crc}
# Random sample of genes (10 random samples taken) for the non TCGA dataset.
# We will take the average of these 10 samplings. We will select the same number
# of genes as the number of genes that give our miRNA metric the best performance

non_tcga_cox <- readRDS("~/Documents/Work/PhD_Program/Hong_Lab/Projects/CC-Singlecell/Data/Reproducible-results/Rmarkdown/Outputs/non_tcga_cox.rds")

cox_df <- non_tcga_cox

random_non_tcga1 <- sample(colnames(cox_df), size = 350)
random_non_tcga2 <- sample(colnames(cox_df), size = 350)
random_non_tcga3 <- sample(colnames(cox_df), size = 350)
random_non_tcga4 <- sample(colnames(cox_df), size = 350)
random_non_tcga5 <- sample(colnames(cox_df), size = 350)
random_non_tcga6 <- sample(colnames(cox_df), size = 350)
random_non_tcga7 <- sample(colnames(cox_df), size = 350)
random_non_tcga8 <- sample(colnames(cox_df), size = 350)
random_non_tcga9 <- sample(colnames(cox_df), size = 350)
random_non_tcga10 <- sample(colnames(cox_df), size = 350)

# Saving the random results for reproducibility
# write.csv(random_non_tcga1, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes1.csv")
# write.csv(random_non_tcga2, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes2.csv")
# write.csv(random_non_tcga3, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes3.csv")
# write.csv(random_non_tcga4, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes4.csv")
# write.csv(random_non_tcga5, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes5.csv")
# write.csv(random_non_tcga6, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes6.csv")
# write.csv(random_non_tcga7, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes7.csv")
# write.csv(random_non_tcga8, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes8.csv")
# write.csv(random_non_tcga9, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes9.csv")
# write.csv(random_non_tcga10, "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes10.csv")

# Putting all the gene vectors together in a list to loop over for cox model
random_non_tcga_gene_list <- list(
  random_non_tcga1, random_non_tcga2, random_non_tcga3,
  random_non_tcga4, random_non_tcga5, random_non_tcga6,
  random_non_tcga7, random_non_tcga8, random_non_tcga9,
  random_non_tcga10
)

# saveRDS(random_non_tcga_gene_list, file = "Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_genes.rds")
```

```{r ideal random gene performance on non tcga crc}
all_random_gene_cindices <- seq(1, 10, 1)

for (rg in all_random_gene_cindices) {
  current_genes <- random_non_tcga_gene_list[[rg]]
  cox_model <- cox_model_fitter(
    my.seed = 1,
    cox.predictors = current_genes,
    cox.df = cox_df,
    gene.num = 350,
    tumor.stage = FALSE,
    tumor.n = FALSE,
    tumor.m = FALSE,
    n.folds = 10,
    my.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/random_non_tcga_coefs_for_random_genes.csv")
  )

  # Getting the top concordance index from the cross validation and then rounding
  # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
  # the all_random_gene_cindices list with the result
  top_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
  all_random_gene_cindices[rg] <- top_cindex
}

write.csv(
  all_random_gene_cindices,
  "Data/Reproducible-results/Rmarkdown/Outputs/non_tcga_crc_random_genes.csv"
)

mean_random <- mean(all_random_gene_cindices)
```

## Ideal CC Singlecell MS performance was: 0.7502
```{r ideal cc singlecell ms performance on non tcga crc}
library(tidyverse)
library(glmnet)


source("cox_model.R")
source("model_optimizer.R")


combo_used <- c(
  "1000_1010", "1000_910", "1000_810", "1000_710",
  "1000_610", "1000_510", "1000_410", "1000_310",
  "1000_210", "1000_110", "1000_10"
)

gene_sizes <- seq.int(100, 3000, 50)
ms_cindices <- rep(0, length(gene_sizes))

sde_genes <- readRDS("sde_colon_and_rectal_cancer.rds")
non_tcga_cox <- readRDS("non_tcga_cox.rds")
all_combos_best <- rep(0, length(combo_used))

alpha_value <- 1.0

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "non_tcga_crc"

for (gs in gene_sizes) {
  top_cindices <- c()
  top_weight_cindex <- rep(0, length(combo_used))
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    combo_counter <- 1
    print("Counter started")
    print(paste0("Counter is: ", counter))
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-READ/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna_genes <- mirna.ranking
    optimized <- two_weight_optimizer(
      first.metric = mirna_genes,
      second.metric = sde_genes,
      my.filename = paste0("Optimization_ccs_ms_non_tcga_", c, ".rds")
    )

    print(paste0("Combo ", c, " is optimized"))
    for (ms in optimized[1:11]) {
      print(paste0("We are finding the optimal gene size for a weighting in combo ", c))

      print(paste0(gs, " is the gene size"))
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = 1,
        cox.predictors = ms,
        cox.df = non_tcga_cox,
        gene.num = gs,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        n.folds = 10,
        my.filename = paste0("non_tcga_crc.csv")
      )
      # Getting the top concordance index from the
      # cross validation and then rounding it to
      # 4 digits to follow cv.glmnet reporting convention.
      # Finally, we update the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]],
        digits = 4
      )
      my_cindices[counter] <- current_cindex

      print("Getting the max cindex of the current glmnet model")
      top_cindex <- max(my_cindices)
      top_index <- which.max(my_cindices)
      print("Getting the index of the max cindex of the current glmnet model")
      top_cindices <- c(top_cindices, top_cindex)
      print(paste0("The top c-index of the current Cox model is: ", top_cindex))
      print(paste0("Its weighting is: ", top_index))
    }
    # Finding the best cindex from the 11 weight combinations of
    # the currrent miRNA-miRNA target pair
    print(paste0("The best c-index for the current combo ", c, " is ", max(top_cindices)))
    top_weight_cindex[counter] <- max(top_cindices)
  }
  # The best c-index of all combos
  all_combos_best[combo_counter] <- max(top_weight_cindex)
  combo_counter <- combo_counter + 1
}
# Binding the gene size and c-index vectors
# together to get the finished data frame
top_cindex <- max(top_weight_cindex)
worst_cindex <- min(top_weight_cindex)
top_cindex_loc <- which.max(top_weight_cindex)
worst_cindex_loc <- which.min(top_weight_cindex)
print(paste0("The best c-index for gene size ", gs, "is ", top_cindex))
print(paste0("The worst c-index for gene size ", gs, "is ", worst_cindex))
print(paste0("The best c-index is found at ", top_cindex_loc))
print(paste0("The worst c-index is found at ", worst_cindex_loc))
cc_single_mms_df <- as.data.frame(cbind(
  gene_sizes, top_cindex,
  top_cindex_loc, worst_cindex, worst_cindex_loc
))
print("Dataframe made")
cc_single_mms_df$method <- rep("CC Singlecell MS", nrow(cc_single_mms_df))
colnames(cc_single_mms_df) <- c(
  "gene_size", "best_cindex",
  "best_cindex_weighting", "worst_cindex", "worst_cindex_weighting", "method"
)
print("Renaming columns")
write.csv(cc_single_mms_df,
  file = "cc_singlecell_ms_test.csv"
)
print("Saving finished csv file")
```

## Ideal CC Singlecell MM performance was 0.7459
```{r ideal performance of cc singlecell mm on non tcga crc}
library(tidyverse)
library(glmnet)


source("cox_model.R")
source("model_optimizer.R")


combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010",
  "800_10", "800_110", "800_210", "800_310", "800_410", "800_510", "800_610", "800_710", "800_810", "800_910", "800_1010",
  "700_10", "700_110", "700_210", "700_310", "700_410", "700_510", "700_610", "700_710", "700_810", "700_910", "700_1010",
  "600_10", "600_110", "600_210", "600_310", "600_410", "600_510", "600_610", "600_710", "600_810", "600_910", "600_1010",
  "500_10", "500_110", "500_210", "500_310", "500_410", "500_510", "500_610", "500_710", "500_810", "500_910", "500_1010",
  "400_10", "400_110", "400_210", "400_310", "400_410", "400_510", "400_610", "400_710", "400_810", "400_910", "400_1010",
  "300_10", "300_110", "300_210", "300_310", "300_410", "300_510", "300_610", "300_710", "300_810", "300_910", "300_1010",
  "200_10", "200_110", "200_210", "200_310", "200_410", "200_510", "200_610", "200_710", "200_810", "200_910", "200_1010",
  "100_10", "100_110", "100_210", "100_310", "100_410", "100_510", "100_610", "100_710", "100_810", "100_910", "100_1010"
)

gene_sizes <- seq(100, 3000, 50)

mad.genes <- readRDS("mad_colon_and_rectal_cancer.rds")
non_tcga_cox <- readRDS("non_tcga_cox.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "non_tcga_crc"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[1:110]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-READ/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_sde_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (ms in mirna_sde_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          cox.predictors = ms,
          cox.df = non_tcga_cox,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          n.folds = 10,
          my.filename = paste0("cc_singlecell_mm_ideal_gene_size_coefs_non_tcga_crc.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_mm_non_tcga_crc_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_mm_non_tcga_crc_df_all_gene_size_data.csv"))
  }
}
```

## Ideal CC Singlecell MMS performance was 0.7911
```{r ideal performance of cc singlecell mms on non tcga crc}
library(tidyverse)
library(glmnet)


source("cox_model.R")
source("model_optimizer.R")


combo_used <- c(
  "1000_1010", "1000_910", "1000_810", "1000_710",
  "1000_610", "1000_510", "1000_410", "1000_310",
  "1000_210", "1000_110", "1000_10"
)

gene_sizes <- 100
ms_cindices <- rep(0, length(gene_sizes))

sde_genes <- readRDS("sde_colon_and_rectal_cancer.rds")
mad_genes <- readRDS("mad_colon_and_rectal_cancer.rds")
non_tcga_cox <- readRDS("non_tcga_cox.rds")
all_combos_best <- rep(0, length(combo_used))

alpha_value <- 1.0

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "non_tcga_crc"

for (gs in gene_sizes) {
  top_cindices <- c()
  top_weight_cindex <- rep(0, length(combo_used))
  for (c in combo_used) {
    my_cindices <- rep(0, 11)
    counter <- 1
    combo_counter <- 1
    print("Counter started")
    print(paste0("Counter is: ", counter))
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-READ/Data/miRNA_inputs/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna_genes <- mirna.ranking
    optimized <- three_weight_optimizer(
      first.metric = mirna_genes,
      second.metric = mad_genes,
      third.metric = mirna_genes,
      my.filename = paste0("Optimization_ccs_ms_non_tcga_", c, ".rds")
    )

    print(paste0("Combo ", c, " is optimized"))
    for (ms in optimized[1:11]) {
      print(paste0("We are finding the optimal gene size for a weighting in combo ", c))

      print(paste0(gs, " is the gene size"))
      cox_model <- cox_model_fitter(
        my.seed = 1,
        my.alpha = 1,
        cox.predictors = ms,
        cox.df = non_tcga_cox,
        gene.num = gs,
        tumor.stage = FALSE,
        tumor.n = FALSE,
        tumor.m = FALSE,
        n.folds = 10,
        my.filename = paste0("non_tcga_crc.csv")
      )
      # Getting the top concordance index from the
      # cross validation and then rounding it to
      # 4 digits to follow cv.glmnet reporting convention.
      # Finally, we update the c_index list with the result
      current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]],
        digits = 4
      )
      my_cindices[counter] <- current_cindex

      print("Getting the max cindex of the current glmnet model")
      top_cindex <- max(my_cindices)
      top_index <- which.max(my_cindices)
      print("Getting the index of the max cindex of the current glmnet model")
      top_cindices <- c(top_cindices, top_cindex)
      print(paste0("The top c-index of the current Cox model is: ", top_cindex))
      print(paste0("Its weighting is: ", top_index))
    }
    # Finding the best cindex from the 11 weight combinations of
    # the currrent miRNA-miRNA target pair
    print(paste0("The best c-index for the current combo ", c, " is ", max(top_cindices)))
    top_weight_cindex[counter] <- max(top_cindices)
  }
  # The best c-index of all combos
  all_combos_best[combo_counter] <- max(top_weight_cindex)
  combo_counter <- combo_counter + 1
}
# Binding the gene size and c-index vectors
# together to get the finished data frame
top_cindex <- max(top_weight_cindex)
worst_cindex <- min(top_weight_cindex)
top_cindex_loc <- which.max(top_weight_cindex)
worst_cindex_loc <- which.min(top_weight_cindex)
print(paste0("The best c-index for gene size ", gs, "is ", top_cindex))
print(paste0("The worst c-index for gene size ", gs, "is ", worst_cindex))
print(paste0("The best c-index is found at ", top_cindex_loc))
print(paste0("The worst c-index is found at ", worst_cindex_loc))
cc_single_mms_df <- as.data.frame(cbind(
  gene_sizes, top_cindex,
  top_cindex_loc, worst_cindex, worst_cindex_loc
))
print("Dataframe made")
cc_single_mms_df$method <- rep("CC Singlecell MMS", nrow(cc_single_mms_df))
colnames(cc_single_mms_df) <- c(
  "gene_size", "best_cindex",
  "best_cindex_weighting", "worst_cindex", "worst_cindex_weighting", "method"
)
print("Renaming columns")
write.csv(cc_single_mms_df,
  file = "cc_singlecell_mms_test.csv"
)
print("Saving finished csv file")
```

```{r ideal cc singlecell mm gene size on non-tcga-crc}
combo_used <- c(
  "1000_10", "1000_110", "1000_210", "1000_310", "1000_410", "1000_510", "1000_610", "1000_710", "1000_810", "1000_910", "1000_1010",
  "900_10", "900_110", "900_210", "900_310", "900_410", "900_510", "900_610", "900_710", "900_810", "900_910", "900_1010"
)

gene_sizes <- seq(100, 3000, 50)
mad.genes <- readRDS("sde_colon_and_rectal_cancer.rds")
cox_df <- readRDS("read_df_finished.rds")

alpha_value <- c(0.0, 0.5, 1.0)

mirna_used <- gsub(combo_used[1], pattern = "_10", replacement = "")
data_set <- "non_tcga_crc"


for (a in alpha_value[3]) {
  top_cindices <- c()
  for (c in combo_used[1:22]) {
    my_cindices <- rep(0, 11)
    counter <- 1
    load(file = paste0("/home/awillems/Projects/CC_Singlecell/TCGA-COAD/Data/Mirna/Inputs/", c, "_targets.RData"), verbose = TRUE)
    mirna.genes <- mirna.ranking
    mirna_mad_optimized <- two_weight_optimizer(
      first.metric = mirna.genes,
      second.metric = mad.genes,
      my.filename = paste0("Optimizations/Optimization_", c, "_targets_cc_singlecell_mm_", a, "_alpha_", data_set, ".rds")
    )







    for (mm in mirna_mad_optimized[1:11]) {
      for (gs in gene_sizes) {
        cox_model <- cox_model_fitter(
          my.seed = 1,
          my.alpha = a,
          cox.predictors = mm,
          cox.df = cox_df,
          gene.num = gs,
          tumor.stage = FALSE,
          tumor.n = FALSE,
          tumor.m = FALSE,
          n.folds = 10,
          my.filename = paste0("cc_singlecell_mm_ideal_gene_size_coefs_non_tcga_crc.csv")
        )

        # Getting the top concordance index from the cross validation and then rounding
        # it to 4 digits to follow cv.glmnet reporting convention. Finally, we update
        # the c_index list with the result
        current_cindex <- round(cox_model$CV$cvm[cox_model$CV$index[1]], digits = 4)
        my_cindices[counter] <- current_cindex
        counter <- counter + 1
      }

      top_cindex <- max(my_cindices)
      top_index <- which(my_cindices == top_cindex)
      print(top_index)
      print(top_cindex)
      top_index_used <- top_index[1]
      top_cindices <- c(top_cindices, top_cindex)
    }

    write.csv(top_cindices, file = paste0("cc_singlecell_mm_non_tcga_crc_df_ideal_gene_size.csv"))
    write.csv(my_cindices, file = paste0("cc_singlecell_mm_non_tcga_crc_df_all_gene_size_data.csv"))
  }
}
```


## DEsingle non tcga crc: 0.5625
```{r c-index for DEsingle on non-tcga-crc}
# For finding the optimal number of genes----
gene_sizes <- seq(100, 5750, 50)
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

des_results <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/des_results_read.rds")

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = non_tcga_cox,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(des_results)[1:gs],
    n.folds = 10,
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/desingle_non_tcga_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```

## scDD non tcga crc: 0.6364
```{r ideal scDD performance on non tcga crc}
# For finding the optimal number of genes
gene_sizes <- seq(100, 3000, 50)
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

scdd_res <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/scdd_res.rds")

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = non_tcga_cox,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = scdd_res$gene[1:gs],
    n.folds = 10,
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/scdd_non_tcga_coefs.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

which.max(my_cindices)
my_cindices[5]
```

## edgeR non tcga crc: 0.702
```{r finding ideal gene number for edgeR for READ}
# For finding the optimal number of genes
gene_sizes <- seq(100, 1930, 50)
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

finished_edger <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/edgeR_1927_genes.rds")

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = non_tcga_cox,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = rownames(finished_edger)[1:gs],
    n.folds = 10,
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/edger_coefs_non_tcga.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```

## DESeq2 non tcga crc: 0.5667 
```{r deseq2 on non tcga crc}
gene_sizes <- seq(100, 1930, 50)
my_cindices <- rep(0, length(gene_sizes))
counter <- 1

finished_deseq <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/deseq2_top_genes.rds")

for (gs in gene_sizes) {
  current_cox <- cox_model_fitter(
    my.seed = 1, my.alpha = 1, cox.df = non_tcga_cox,
    gene.num = gs, tumor.m = FALSE,
    tumor.n = FALSE, tumor.stage = FALSE,
    cox.predictors = finished_deseq[1:gs],
    n.folds = 10,
    my.filename = "Data/Reproducible-results/Rmarkdown/Outputs/deseq2_coefs_non_tcga.csv"
  )

  my_cindices[counter] <- round(current_cox$CV$cvm[current_cox$CV$index[1]],
    digits = 4
  )

  counter <- counter + 1
}

print(max(my_cindices))
```


## LASSO validation code and files
## Validation that LASSO regularization is indeed best for our data: Doing grid search for COAD at different values of alpha to show that LASSO is best regularization
```{r data for miRNA-miRNA target number grid search for COAD at different values of alpha}
# CC Singlecell MS grid search COAD
# See server_speedup_coad.R for the code


# CC Singlecell MS grid search READ
# See server_speedup_read.R for the code


# Reading in the results files of the CC Singlecell MS COAD grid search
coad_ms_0 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0_cc_singlecell_ms_coad_used_combo_100_1010_index_5_df.csv")
colnames(coad_ms_0) <- c("number", "c_index")
coad_ms_0$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_ms_0$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)

coad_ms_05 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0.5_cc_singlecell_ms_coad_used_combo_100_1010_index_5_df.csv")
colnames(coad_ms_05) <- c("number", "c_index")
coad_ms_05 <- coad_ms_05[89:176, ]
coad_ms_05$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_ms_05$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)

coad_ms_1 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_1_cc_singlecell_ms_coad_used_combo_100_1010_index_5_df.csv")
colnames(coad_ms_1) <- c("number", "c_index")
coad_ms_1 <- coad_ms_1[177:264, ]
coad_ms_1$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_ms_1$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)
```


```{r plots for miRNA-miRNA target number grid search for COAD at alpha 0}
# Plotting the results of the CC Singlecell MS COAD grid search
# Alpha 0
heatmap_coad_ccs_ms_0 <- ggplot(data = coad_ms_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS COAD Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_ms_finished <- heatmap_coad_ccs_ms_0 + scale_fill_viridis_c()
```


```{r saving plots for miRNA-miRNA target number grid search for COAD at alpha 0}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_coad_grid_search_heatmap_alpha_0_figs2.svg",
  plot = print(heatmap_coad_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "cm"
)
```


```{r plots for miRNA-miRNA target number grid search for COAD at alpha 0.5}
# Alpha 0.5
heatmap_coad_ccs_ms_05 <- ggplot(data = coad_ms_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS COAD Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_ms_finished <- heatmap_coad_ccs_ms_05 + scale_fill_viridis_c()
```


```{r saving plots for miRNA-miRNA target number grid search for COAD at alpha 0.5}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_coad_grid_search_heatmap_alpha_05_figs2.svg",
  plot = print(heatmap_coad_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 290, height = 290,
  units = "mm"
)
```


```{r plots for miRNA-miRNA target number grid search for COAD at alpha 1}
# Alpha 1
heatmap_coad_ccs_ms_1 <- ggplot(data = coad_ms_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS COAD Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_ms_finished <- heatmap_coad_ccs_ms_1 + scale_fill_viridis_c()
```


```{r saving plots for miRNA-miRNA target number grid search for COAD at alpha 0.5}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_coad_grid_search_heatmap_alpha_1_figs2.svg",
  plot = print(heatmap_coad_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r data for miRNA-miRNA target number grid search for READ}
# Reading in the results files of the CC Singlecell MS READ grid search
read_ms_0 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0_cc_singlecell_ms_read_used_combo_100_1010_index_8_df copy.csv")
colnames(read_ms_0) <- c("number", "c_index")
read_ms_0$mirna_num <- rep(seq(100, 1000, 100), 11)
read_ms_0$mirna_target <- rep(seq(10, 1010, by = 100), 10)


read_ms_05 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0.5_cc_singlecell_ms_read_used_combo_100_1010_index_8_df.csv")
colnames(read_ms_05) <- c("number", "c_index")
read_ms_05 <- read_ms_05[111:220, ]
read_ms_05$mirna_num <- rep(seq(100, 1000, 100), 11)
read_ms_05$mirna_target <- rep(seq(10, 1010, by = 100), 10)

read_ms_1 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_1_cc_singlecell_ms_read_used_combo_100_1010_index_11_df.csv")
colnames(read_ms_1) <- c("number", "c_index")
read_ms_1 <- read_ms_1[221:330, ]
read_ms_1$mirna_num <- rep(seq(100, 1000, 100), 11)
read_ms_1$mirna_target <- rep(seq(10, 1010, by = 100), 10)
```


```{r plotting for miRNA-miRNA target number grid search for READ at alpha 0}
# Plotting the results of the CC Singlecell MS READ grid search
# READ MS Alpha 0
heatmap_read_ccs_ms_0 <- ggplot(data = read_ms_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS READ Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ms_0_ccs_ms_finished <- heatmap_read_ccs_ms_0 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for READ at alpha 0}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_read_grid_search_heatmap_alpha_0.svg",
  plot = print(heatmap_read_ms_0_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for READ at alpha 0.5}
# READ MS Alpha 0.5
heatmap_read_ccs_ms_05 <- ggplot(data = read_ms_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS READ Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ms_05_ccs_ms_finished <- heatmap_read_ccs_ms_05 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for READ at alpha 0.5}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_read_grid_search_heatmap_alpha_05.svg",
  plot = print(heatmap_read_ms_05_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for READ at alpha 1}
# READ MS Alpha 1
heatmap_read_ccs_ms_1 <- ggplot(data = read_ms_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MS READ Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ms_1_ccs_ms_finished <- heatmap_read_ccs_ms_1 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for READ at alpha 1}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_ms_read_grid_search_heatmap_alpha_1.svg",
  plot = print(heatmap_read_ms_1_ccs_ms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r data for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 0}
# CC Singlecell MM COAD
coad_mm_0 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0_cc_singlecell_mm_coad_used_combo_100_1010_index_5_df.csv")
colnames(coad_mm_0) <- c("number", "c_index")
coad_mm_0$mirna_num <- rep(seq(100, 800, 100), 11)
coad_mm_0$mirna_target <- rep(seq(10, 1010, by = 100), 8)

coad_mm_05 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0.5_cc_singlecell_mm_coad_used_combo_100_1010_index_10_df.csv")
colnames(coad_mm_05) <- c("number", "c_index")
coad_mm_05$mirna_num <- rep(seq(100, 800, 100), 11)
coad_mm_05$mirna_target <- rep(seq(10, 1010, by = 100), 8)


coad_mm_1 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_1_cc_singlecell_mm_coad_used_combo_100_1010_index_10_df.csv")
colnames(coad_mm_1) <- c("number", "c_index")
coad_mm_1$mirna_num <- rep(seq(100, 800, 100), 11)
coad_mm_1$mirna_target <- rep(seq(10, 1010, by = 100), 8)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 0}
# Alpha 0
heatmap_coad_ccs_mm_0 <- ggplot(data = coad_mm_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM COAD Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mm_finished <- heatmap_coad_ccs_mm_0 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 0}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mm_coad_grid_search_heatmap_alpha_0.svg",
  plot = print(heatmap_coad_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 0.5}
# Alpha 0.5
heatmap_coad_ccs_mm_05 <- ggplot(data = coad_mm_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM COAD Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans", margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mm_finished <- heatmap_coad_ccs_mm_05 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 0.5}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mm_coad_grid_search_heatmap_alpha_05.svg",
  plot = print(heatmap_coad_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 1}
# Alpha 1
heatmap_coad_ccs_mm_1 <- ggplot(data = coad_mm_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM COAD Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans", margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mm_finished <- heatmap_coad_ccs_mm_1 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM COAD at alpha 1}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mm_coad_grid_search_heatmap_alpha_1.svg",
  plot = print(heatmap_coad_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 34, height = 34,
  units = "mm"
)
```


```{r data for miRNA-miRNA target number grid search for CC Singlecell MM READ}
# CC Singlecell MM READ
read_mm_0 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0_cc_singlecell_mm_read_used_combo_100_1010_index_1_df.csv")
colnames(read_mm_0) <- c("number", "c_index")
read_mm_0$mirna_num <- rep(seq(100, 1000, 100), 11)
read_mm_0$mirna_target <- rep(seq(10, 1010, by = 100), 10)

read_mm_05 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0.5_cc_singlecell_mm_read_used_combo_100_1010_index_7_df.csv")
colnames(read_mm_05) <- c("number", "c_index")
read_mm_05$mirna_num <- rep(seq(100, 1000, 100), 11)
read_mm_05$mirna_target <- rep(seq(10, 1010, by = 100), 10)

read_mm_1 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_1_cc_singlecell_mm_read_used_combo_100_1010_index_10_df.csv")
colnames(read_mm_1) <- c("number", "c_index")
read_mm_1$mirna_num <- rep(seq(100, 1000, 100), 11)
read_mm_1$mirna_target <- rep(seq(10, 1010, by = 100), 10)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 0}
# Alpha 0
heatmap_read_ccs_mm_0 <- ggplot(data = read_mm_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM READ Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans", margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mm_finished <- heatmap_read_ccs_mm_0 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 0}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mm_read_grid_search_heatmap_alpha_0.svg",
  plot = print(heatmap_read_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 0.5}
# Alpha 0.5
heatmap_read_ccs_mm_05 <- ggplot(data = read_mm_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM READ Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.0, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mm_finished <- heatmap_read_ccs_mm_05 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 0.5}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mm_read_grid_search_heatmap_alpha_05.svg",
  plot = print(heatmap_read_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 1}
# Alpha 1
heatmap_read_ccs_mm_1 <- ggplot(data = read_mm_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MM READ Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.0, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mm_finished <- heatmap_read_ccs_mm_1 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MM READ at alpha 1}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/cc_singlecell_mm_read_grid_search_heatmap_alpha_1.svg",
  plot = print(heatmap_read_ccs_mm_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r data for miRNA-miRNA target number grid search for CC Singlecell MMS COAD}
# CC Singlecell MMS COAD
coad_mms_0 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0_cc_singlecell_mms_coad_used_combo_100_1010_index_86_df.csv")
colnames(coad_mms_0) <- c("number", "c_index")
coad_mms_0$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_mms_0$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)

coad_mms_05 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_0.5_cc_singlecell_mms_coad_used_combo_100_1010_index_17_df.csv")
colnames(coad_mms_05) <- c("number", "c_index")
coad_mms_05$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_mms_05$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)


coad_mms_1 <- read.csv("Data/Reproducible-results/Data/Outputs/COAD/top_cindices_alpha_1_cc_singlecell_mms_coad_used_combo_100_1010_index_17_df.csv")
colnames(coad_mms_1) <- c("number", "c_index")
coad_mms_1$mirna_num <- rep(seq(100, 800, 100), each = 11)
coad_mms_1$mirna_target <- rep(seq(10, 1010, by = 100), times = 8)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 0}
# Alpha 0
heatmap_coad_ccs_mms_0 <- ggplot(data = coad_mms_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS COAD Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, face = "bold"),
    axis.title.y = element_text(size = 40, face = "bold"),
    axis.text.x = element_text(size = 30),
    axis.text.y = element_text(size = 40),
    legend.text = element_text(size = 25),
    legend.title = element_text(size = 40),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mms_finished <- heatmap_coad_ccs_mms_0 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 0}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_coad_grid_search_heatmap_alpha_0_figs2.svg",
  plot = print(heatmap_coad_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 290, height = 290,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 0.5}
# Alpha 0.5
heatmap_coad_ccs_mms_05 <- ggplot(data = coad_mms_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS COAD Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, face = "bold"),
    axis.title.y = element_text(size = 40, face = "bold"),
    axis.text.x = element_text(size = 30),
    axis.text.y = element_text(size = 40),
    legend.text = element_text(size = 25),
    legend.title = element_text(size = 40, margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mms_finished <- heatmap_coad_ccs_mms_05 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 0.5}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_coad_grid_search_heatmap_alpha_05_figs2.svg",
  plot = print(heatmap_coad_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 290, height = 290,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 1}
# Alpha 1
heatmap_coad_ccs_mms_1 <- ggplot(data = coad_mms_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS COAD Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    text = element_text(family = "Arial"),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, face = "bold"),
    axis.title.y = element_text(size = 40, face = "bold"),
    axis.text.x = element_text(size = 30),
    axis.text.y = element_text(size = 40),
    legend.text = element_text(size = 25),
    legend.title = element_text(size = 40, margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_coad_ccs_mms_finished <- heatmap_coad_ccs_mms_1 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS COAD at alpha 1}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_coad_grid_search_heatmap_alpha_1_figs2.svg",
  plot = print(heatmap_coad_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 320, height = 320,
  units = "mm"
)
```


```{r data for miRNA-miRNA target number grid search for CC Singlecell MMS READ}
# CC Singlecell MMS READ
read_mms_0 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0_cc_singlecell_mms_read_used_combo_900_1010_index_56_df.csv")
colnames(read_mms_0) <- c("number", "c_index")
read_mms_0$mirna_num <- rep(seq(100, 1000, 100), each = 11)
read_mms_0$mirna_target <- rep(seq(10, 1010, by = 100), times = 10)

read_mms_05 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_0.5_cc_singlecell_mms_read_used_combo_100_1010_index_7_df.csv")
colnames(read_mms_05) <- c("number", "c_index")
read_mms_05$mirna_num <- rep(seq(100, 1000, 100), 11)
read_mms_05$mirna_target <- rep(seq(10, 1010, by = 100), 10)

read_mms_1 <- read.csv("Data/Reproducible-results/Data/Outputs/READ/top_cindices_alpha_1_cc_singlecell_mms_read_used_combo_100_1010_index_10_df.csv")
colnames(read_mms_1) <- c("number", "c_index")
read_mms_1$mirna_num <- rep(seq(100, 1000, 100), 11)
read_mms_1$mirna_target <- rep(seq(10, 1010, by = 100), 10)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 0}
# Alpha 0
heatmap_read_ccs_mms_0 <- ggplot(data = read_mms_0, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS READ Alpha = 0",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans", margin = margin(0, 20, 0, 0)),
    legend.position = "bottom",
    legend.key.width = unit(2.5, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mms_finished <- heatmap_read_ccs_mms_0 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 0}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_read_grid_search_heatmap_alpha_0.svg",
  plot = print(heatmap_read_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 0.5}
# Alpha 0.5
heatmap_read_ccs_mms_05 <- ggplot(data = read_mms_05, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS READ Alpha = 0.5",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.0, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mms_finished <- heatmap_read_ccs_mms_05 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 0.5}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_read_grid_search_heatmap_alpha_05.svg",
  plot = print(heatmap_read_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```


```{r plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 1}
# Alpha 1
heatmap_read_ccs_mms_1 <- ggplot(data = read_mms_1, aes(x = mirna_num, y = mirna_target, fill = c_index)) +
  geom_tile() +
  scale_fill_gradient(low = "white", high = "red") +
  geom_text(aes(label = round(c_index, 4)), color = "white") +
  coord_fixed() +
  labs(
    x = "# of miRNAs",
    y = "# of miRNA Targets",
    title = "CC Singlecell MMS READ Alpha = 1",
    fill = "Concordance Index"
  ) +
  theme(
    panel.background = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold", size = 40),
    axis.title.x = element_text(size = 40, family = "sans", face = "bold"),
    axis.title.y = element_text(size = 40, family = "sans", face = "bold"),
    axis.text.x = element_text(size = 30, family = "sans"),
    axis.text.y = element_text(size = 40, family = "sans"),
    legend.text = element_text(size = 25, family = "sans"),
    legend.title = element_text(size = 40, family = "sans"),
    legend.position = "bottom",
    legend.key.width = unit(2.0, "cm")
  )

# Changing to color-blind friendly palette
heatmap_read_ccs_mms_finished <- heatmap_read_ccs_mms_1 + scale_fill_viridis_c()
```


```{r saving plotting for miRNA-miRNA target number grid search for CC Singlecell MMS READ at alpha 1}
# Now saving the heat map
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/cc_singlecell_mms_read_grid_search_heatmap_alpha_1.svg",
  plot = print(heatmap_read_ccs_mms_finished, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 32, height = 32,
  units = "mm"
)
```

## Response to Reviewer 1
```{r distribution of individual coad metrics}
mad_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
load(file = "Data/Reproducible-results/Rmarkdown/Data/700_810_targets.RData", verbose = TRUE)
mirna_genes <- mirna.ranking

mad_genes_df <- as.data.frame(mad_genes)
sde_genes_df <- as.data.frame(sde_genes)
mirna_genes_df <- as.data.frame(mirna_genes)

p1 <- ggplot(data = mad_genes_df, aes(x = mad_genes)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("MAD") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))

p2 <- ggplot(data = sde_genes_df, aes(x = sde_genes)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("SDE") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))

p3 <- ggplot(data = mirna_genes_df, aes(x = mirna_genes)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("miRNA") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))
```


```{r distribution of individual read metrics}
mad_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
load(file = "Data/Reproducible-results/Rmarkdown/Data/300_310_targets.RData", verbose = TRUE)
mirna_genes <- mirna.ranking

mad_genes_df <- as.data.frame(mad_genes)
sde_genes_df <- as.data.frame(sde_genes)
mirna_genes_df <- as.data.frame(mirna_genes)

p4 <- ggplot(data = mad_genes_df, aes(x = mad_genes)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("MAD") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))

p5 <- ggplot(data = sde_genes_df, aes(x = sde_genes)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("SDE") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))

p6 <- ggplot(data = mirna_genes_df, aes(x = mirna_genes)) +
  geom_histogram(bins = 15) +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5),
    plot.margin = unit(c(0.8, 0.8, 0.8, 0.8), units = "cm")
  ) +
  ggtitle("miRNA") +
  xlab("Scores") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number(), limits = c(0, 0.00020))
```

```{r saving metric distributions}
my_plots <- list(p1, p2, p3, p4, p5, p6)
for (p in 1:length(my_plots)) {
  ggsave(
    filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/", p, "_distribution.png"),
    plot = print(my_plots[[p]], newpage = FALSE),
    device = "png", dpi = 600,
    width = 7.8, height = 7.8,
    units = "in"
  )
}
```

```{r pseudotime generation for correlation analysis of emt genes}
# Generating the different genes based pseudotime progression with Monocle3
# Based on the expression of the each graph we select the point that is
# approximately the lowest level of expression
# as the root of our pseudotime as this matches the gradient of gene expression
# level
tian_emt_genes <- read.csv("Data/Reproducible-results/Rmarkdown/emt_genes_tian.csv")
my_genes <- tian_emt_genes$Gene
common_emt_genes <- intersect(tian_emt_genes$Gene, my_sc_df_names$gene_short_name)
tian_emt_genes_e <- filter(tian_emt_genes, Gene %in% common_emt_genes & Annotation == "E")
tian_emt_genes_m <- filter(tian_emt_genes, Gene %in% common_emt_genes & Annotation == "M")
random_genes_e <- tian_emt_genes_e %>% slice_sample(n = 80)
random_genes_m <- tian_emt_genes_m %>% slice_sample(n = 80)
# E gene loop
for (g in random_genes_e$Gene[1:nrow(random_genes_e)]) {
  cds_output <- cell_dataset_builder(
    vim.genes = random_genes_e$Gene[g],
    cell.data = my_sc_df,
    my.monocle.plot.dpi = 600,
    cell.meta = my_sc_df_names,
    plot.cells = FALSE,
    use.function = FALSE,
    my.root = "Y_4",
    norm.flag = "log",
    my.cds.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/cds_ouput_all_tian_e_genes_", g, ".rds"),
    my.monocle.graph = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_all_tian_e_genes_", g, ".svg"),
    my.monocle.graph.genes = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_genes_all_tian_e_genes_", g, ".svg"),
    my.pt.data.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_data_all_tian_e_genes_", g, ".csv")
  )
}

# M gene loop
for (g in random_genes_m$Gene[1:nrow(random_genes_m)]) {
  cds_output <- cell_dataset_builder(
    vim.genes = random_genes_m$Gene[g],
    cell.data = my_sc_df,
    cell.meta = my_sc_df_names,
    plot.cells = FALSE,
    use.function = FALSE,
    my.root = "Y_7",
    my.monocle.plot.dpi = 600,
    norm.flag = "log",
    my.cds.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/cds_ouput_all_tian_m_genes_", g, ".rds"),
    my.monocle.graph = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_all_tian_m_genes_", g, ".svg"),
    my.monocle.graph.genes = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_genes_all_tian_m_genes_", g, ".svg"),
    my.pt.data.filename = paste0("Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_data_all_tian_m_genes_", g, ".csv")
  )
}
```


```{r e gene expression levels}
my_sc_df <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data.rds")
my_sc_df_names <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/denoised-colon-and-rectal-single-cell-data-gene-names.rds")
random_genes_e <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/tian_emt_genes_e_set_randomly_selected.csv")
cds_output <- cell_dataset_builder(
  vim.genes = random_genes_e$Gene[1:30],
  cell.data = my_sc_df,
  cell.meta = my_sc_df_names,
  my.monocle.plot.dpi = 600,
  my.root = "Y_7",
  point.size = 2.5,
  use.function = FALSE,
  save.gene.expr = TRUE,
  norm.flag = "log",
  my.cds.filename = "Data/Reproducible-results/Rmarkdown/Outputs/cds_ouput_test5.rds",
  my.monocle.graph = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_test5.svg",
  my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/30_e_genes_expression_profile_updated_600_dpi.svg",
  my.pt.data.filename = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_data_test5.csv"
)
cds_output$`Cell Progression Graph`
```

```{r m gene expression plots}
random_genes_m <- read.csv("Data/Reproducible-results/Rmarkdown/Outputs/tian_emt_genes_m_set_randomly_selected.csv")
cds_output <- cell_dataset_builder(
  vim.genes = random_genes_m$Gene[1:30],
  # vim.genes = c("VIM", "VIMP", "CDH1", "ZEB1"),
  cell.data = my_sc_df,
  cell.meta = my_sc_df_names,
  my.root = "Y_7",
  my.monocle.plot.dpi = 600,
  point.size = 2.5,
  use.function = FALSE,
  save.gene.expr = TRUE,
  norm.flag = "log",
  my.cds.filename = "Data/Reproducible-results/Rmarkdown/Outputs/cds_ouput_test5.rds",
  my.monocle.graph = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_test5.svg",
  # my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/hallmark_emt_genes_expression_profile_updated_600_dpi.svg",
  my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/30_m_emt_genes_expression_profile_updated_600_dpi.svg",
  my.pt.data.filename = "Data/Reproducible-results/Rmarkdown/Outputs/vim_pseudotime_data_test5.csv"
)
cds_output$`Cell Progression Graph`
```


```{r corr analysis function}
do_corr_analysis <- function(cds_output, save_mat = TRUE) {
  cell_df <- cds_output$`Cell Progression Graph`$data$value
  cell_df <- as.data.frame(cell_df)
  colnames(cell_df)[1] <- "Expr"
  rownames(cell_df) <- colnames(my_sc_df)
  pt_df <- cds_output$`PT Graph`$data$cell_color
  pt_df <- as.data.frame(pt_df)
  colnames(pt_df)[1] <- "PT"
  rownames(pt_df) <- colnames(my_sc_df)
  my_corr_df <- bind_cols(cell_df, pt_df)
  my_corr_df$PT <- format(my_corr_df$PT, scientific = FALSE)
  my_corr_df <- apply(my_corr_df, c(1, 2), as.numeric)
  M <- cor(my_corr_df, use = "complete.obs", method = "spearman")
  M
  if (save_mat == TRUE) {
    write.csv(x = M, file = paste0("Data/Reproducible-results/Rmarkdown/Outputs/m-gene-corr-matricies/", g, ".csv"))
  }
  return(M)
}
```


```{r vim and other canidate markers pseudotime (using vim pseudotime root for e gene analysis)}
for (g in tian_emt_genes_e$Gene) {
  cds_output <- cell_dataset_builder(
    vim.genes = c(g),
    cell.data = my_sc_df,
    cell.meta = my_sc_df_names,
    my.root = "Y_7",
    point.size = 2.5,
    use.function = FALSE,
    norm.flag = "log",
    my.cds.filename = "Data/Reproducible-results/Rmarkdown/Outputs/e-gene-res/cds_ouputput_e_genes_only_with_vim_root.rds",
    my.monocle.graph = "Data/Reproducible-results/Rmarkdown/Outputs/e-gene-res/e-gene-plots/vim_pseudotime_e_genes_only_with_vim_root.svg",
    my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/e-gene-res/e-gene-plots/vim_genes_e_genes_only_with_vim_root.svg",
    my.pt.data.filename = "Data/Reproducible-results/Rmarkdown/Outputs/e-gene-res/vim_pseudotime_data_e_genes_only_with_vim_root.csv"
  )
  do_corr_analysis(cds_output = cds_output, save_mat = TRUE)
}
```


```{r generating correlations for histogram of all e genes}
e_files <- list.files(path = "Data/Reproducible-results/Rmarkdown/Outputs/e-genes-corr-matricies/")
all_cors <- c(rep(0, length(tian_emt_genes_e$Gene)))
counter <- 1
for (f in e_files) {
  current_file <- read.csv(paste0("Data/Reproducible-results/Rmarkdown/Outputs/e-genes-corr-matricies/", f))
  current_corr <- current_file$Expr[2]
  all_cors[counter] <- current_corr
  counter <- counter + 1
}
```


```{r histogram of all e genes}
all_cors <- as.data.frame(all_cors)

e_hist <- ggplot(data = all_cors, aes(x = all_cors)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5)
  ) +
  ggtitle("E Genes") +
  xlab("Spearman Correlations") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number())

e_hist
```

```{r saving e genes histogram}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/e_correlation_histogram.svg",
  plot = print(e_hist, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```
```{r vim and other canidate markers pseudotime (using vim pseudotime root for m gene analysis)}
for (g in tian_emt_genes_m$Gene) {
  cds_output <- cell_dataset_builder(
    vim.genes = c(g),
    cell.data = my_sc_df,
    cell.meta = my_sc_df_names,
    my.root = "Y_7",
    point.size = 2.5,
    use.function = FALSE,
    norm.flag = "log",
    my.cds.filename = "Data/Reproducible-results/Rmarkdown/Outputs/m-gene-res/cds_ouputput_m_genes_only_with_vim_root.rds",
    my.monocle.graph = "Data/Reproducible-results/Rmarkdown/Outputs/m-gene-res/m-gene-plots/vim_pseudotime_m_genes_only_with_vim_root.svg",
    my.monocle.graph.genes = "Data/Reproducible-results/Rmarkdown/Outputs/m-gene-res/m-gene-plots/vim_genes_m_genes_only_with_vim_root.svg",
    my.pt.data.filename = "Data/Reproducible-results/Rmarkdown/Outputs/m-gene-res/vim_pseudotime_data_m_genes_only_with_vim_root.csv"
  )
  do_corr_analysis(cds_output = cds_output, save_mat = TRUE)
}
```



```{r generating correlations for histogram of all m genes}
m_files <- list.files(path = "Data/Reproducible-results/Rmarkdown/Outputs/m-gene-corr-matricies/")
all_cors <- c(rep(0, length(tian_emt_genes_m$Gene)))
counter <- 1
for (f in m_files) {
  current_file <- read.csv(paste0("Data/Reproducible-results/Rmarkdown/Outputs/m-gene-corr-matricies/", f))
  current_corr <- current_file$Expr[2]
  all_cors[counter] <- current_corr
  counter <- counter + 1
}
```


```{r histogram of all m genes}
all_cors <- as.data.frame(all_cors)

m_hist <- ggplot(data = all_cors, aes(x = all_cors)) +
  geom_histogram() +
  theme(
    text = element_text(family = "Arial"),
    panel.background = element_blank(),
    axis.title = element_text(size = 32, face = "bold"),
    axis.text = element_text(size = 12),
    plot.title = element_text(size = 32, face = "bold", hjust = 0.5)
  ) +
  ggtitle("M Genes") +
  xlab("Spearman Correlations") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0), labels = label_number())

m_hist
```

```{r saving m genes histogram}
ggsave(
  filename = "Data/Reproducible-results/Rmarkdown/Outputs/m_correlation_histogram.svg",
  plot = print(m_hist, newpage = FALSE),
  device = "svg", dpi = 600,
  width = 7.8, height = 7.8,
  units = "in"
)
```

```{r test of metric integration}
mad_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/mad_colon_and_rectal_cancer.rds")
sde_genes <- readRDS("Data/Reproducible-results/Rmarkdown/Outputs/sde_colon_and_rectal_cancer.rds")
load(file = "Data/Reproducible-results/Rmarkdown/Data/700_810_targets.RData", verbose = TRUE)
mirna_genes <- mirna.ranking
```
